{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b40cead4",
   "metadata": {},
   "source": [
    "# Library Import and Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8ae625f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "85ccf614",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: relative_transcript_abundance\n",
    "\n",
    "purpose: calculate relative transcript abundance\n",
    "\n",
    "input: a dataframe with a ref_gene_id column identifying the transcript gene of origin and a cov columns with \n",
    "the coverage for the transcripts.\n",
    "\n",
    "output: the same dataframe with median relative abundance and mean gene CPM columns added\n",
    "'''\n",
    "\n",
    "def relative_transcript_abundance(df):\n",
    "    \n",
    "    first=True\n",
    "    \n",
    "    dff = df.copy()\n",
    "\n",
    "    # for each sample's CPM column, calculate the relative abundance of the transcript in the gene and create a new column with that value\n",
    "    for col in dff.filter(regex='bam_CPM').columns:\n",
    "        \n",
    "        col_gene_name = col.split(\"_CP\")[0] + \"_total_gene_CPM\"\n",
    "        col_relative_abundance = col.split(\"_CP\")[0] + \"_relative_abundance\"\n",
    "    \n",
    "        dff_sums = dff[[\"gene_id\", col]].groupby(\"gene_id\").sum()\n",
    "\n",
    "        dff_sums[col_gene_name] = dff_sums[col].copy()\n",
    "\n",
    "        dff_sums.drop(columns=col, inplace=True)\n",
    "\n",
    "        if first:\n",
    "            merged_dff = pd.merge(dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "            merged_dff[col_relative_abundance] = ((merged_dff[col]/merged_dff[col_gene_name]) * 100)\n",
    "            \n",
    "        else:\n",
    "            merged_dff = pd.merge(merged_dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "            merged_dff[col_relative_abundance] = ((merged_dff[col]/merged_dff[col_gene_name]) * 100)\n",
    "        \n",
    "        first=False\n",
    "        \n",
    "    for col_count in merged_dff.filter(regex='bam_count').columns:\n",
    "        \n",
    "        col_gene_name = col_count.split(\"_count\")[0] + \"_total_gene_counts\"\n",
    "        \n",
    "        dff_sums = merged_dff[[\"gene_id\", col_count]].groupby(\"gene_id\").sum()\n",
    "        dff_sums[col_gene_name] = dff_sums[col_count].copy()\n",
    "    \n",
    "        \n",
    "        dff_sums.drop(columns=col_count, inplace=True)\n",
    "        \n",
    "        merged_dff = pd.merge(merged_dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "        \n",
    "    merged_dff.fillna(value=0, inplace=True)\n",
    "    \n",
    "    # calculate the median relative abundance for all the samples and drop the sample specific relative abundance columns\n",
    "    rel_ab_col = merged_dff.filter(regex='bam_relative_abundance').columns\n",
    "    merged_dff['median_relative_abundance'] = merged_dff[rel_ab_col].copy().median(axis=1)\n",
    "    merged_dff.drop(columns = rel_ab_col, inplace = True)\n",
    "    \n",
    "    # calculate the median gene CPM and drop sample specific gene CPM\n",
    "    gene_cpm_col = merged_dff.filter(regex='bam_total_gene_CPM').columns\n",
    "    merged_dff['median_gene_cpm'] = merged_dff[gene_cpm_col].copy().median(axis=1)\n",
    "    merged_dff.drop(columns = gene_cpm_col, inplace = True)\n",
    "        \n",
    "    return merged_dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a6446e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: fix_column_names\n",
    "\n",
    "purpose: Fixing the column names, making them smaller, informative, and consistent\n",
    "\n",
    "input: The raw counts dataframe for either genes or transcripts \n",
    "\n",
    "output: Same dataframe with improved column names\n",
    "'''\n",
    "\n",
    "def fix_column_names(df, is_gene=False):\n",
    "    \n",
    "    ## Check if this is a gene counts object\n",
    "    if is_gene:\n",
    "        \n",
    "        ## Get count column names and create list of new column names\n",
    "        count_columns = df.columns.tolist()\n",
    "        list_new_names = [\"gene_id\"]\n",
    "        \n",
    "        ## gene_id comes in as index for gene counts data, make it into the first column instead\n",
    "        df[\"gene_id\"] = df.index\n",
    "        cols = list(df.columns)\n",
    "        cols = [cols[-1]] + cols[:-1]\n",
    "        df = df[cols]\n",
    "        df.reset_index(inplace=True, drop=True)\n",
    "    \n",
    "    ## If it is a transcript dataset\n",
    "    else:\n",
    "        ## Set count columns and create list of new names\n",
    "        count_columns = df.columns[2:].tolist()\n",
    "        list_new_names = [ \"transcript_id\", \"gene_id\"]\n",
    "    \n",
    "    ## Fix names one by one and add to list of new names\n",
    "    for col in count_columns:\n",
    "        col = col.split(\"_mapped\")[0] + \"_counts\"\n",
    "        list_new_names.append(col)\n",
    "    \n",
    "    ## Rename columns\n",
    "    df.columns = list_new_names\n",
    "    \n",
    "    return df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "143176a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_df_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True, is_transcript=False, is_prot=False):\n",
    "\n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split(\"ne_id \\\"\", expand=True)[1].str.split('\\\";', expand=True)[0]\n",
    "        \n",
    "        ## Get gene names\n",
    "        df[\"gene_name\"] = df[\"other\"].str.split(\"gene_name \\\"\", expand=True)[1].str.split('\\\";', expand=True)[0]\n",
    "        \n",
    "        ## Get get transcript biotype\n",
    "        df[\"gene_biotype\"] = df[\"other\"].str.split('gene_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "        \n",
    "        ## If is transcript get transcript id and transcript biotype\n",
    "        if is_transcript:\n",
    "            df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            df[\"transcript_biotype\"] = df[\"other\"].str.split('transcript_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            \n",
    "            ## If is prot get protein_id\n",
    "            if is_prot:\n",
    "                df[\"protein_id\"] = df[\"other\"].str.split('protein_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "                df[\"ccds_id\"] = df[\"other\"].str.split('ccds_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "                df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in df.columns:\n",
    "        df.loc[df[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1117a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: calculate_cpm\n",
    "\n",
    "purpose: Calculate CPM for the each sample given\n",
    "\n",
    "input: Counts dataset\n",
    "\n",
    "output: Counts dataset with CPM columns as well\n",
    "'''\n",
    "\n",
    "def calculate_cpm(df, is_gene=False):\n",
    "\n",
    "    ## Set count columns if dataframe is gene counts\n",
    "    if is_gene:\n",
    "        count_columns = df.columns[1:].tolist()\n",
    "    \n",
    "    ## Set count columns if dataframe is transcript counts\n",
    "    else:\n",
    "        count_columns = df.columns[2:].tolist()\n",
    "\n",
    "    ## Loop through counts columns to calculate CPM and add to the dataframe\n",
    "    for col in count_columns:\n",
    "        \n",
    "        df[col] = round(df[col], 2)\n",
    "        cpm_name = col.replace(\"_counts\", \"_CPM\")\n",
    "        df[cpm_name] = round(((df[col]/(df[col].sum())) * 1000000), 2)\n",
    "    \n",
    "    return df  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "28b9546f",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: get_samp_id\n",
    "\n",
    "purpose: \n",
    "\n",
    "input: sample string name\n",
    "\n",
    "output: sample string name with bam appended\n",
    "'''\n",
    "\n",
    "\n",
    "def get_samp_id(s):\n",
    "    \n",
    "    name_array = s.split('.')\n",
    "    last_element = name_array[-1].split('_')[0]\n",
    "    string_1 = '-'.join(name_array[:-1])\n",
    "    string_2 = f\"{last_element}.bam\"\n",
    "    return f\"{string_1}.{string_2}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "21852d06",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues_to_use = [\"Brain - Cerebellar Hemisphere\",\n",
    "                   \"Brain - Frontal Cortex (BA9)\",\n",
    "                   \"Brain - Putamen (basal ganglia)\",\n",
    "                   \"Cells - Cultured fibroblasts\",\n",
    "                   \"Heart - Atrial Appendage\",\n",
    "                   \"Heart - Left Ventricle\",\n",
    "                   \"Liver\",\n",
    "                   \"Lung\",\n",
    "                   \"Muscle - Skeletal\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19eb60d8",
   "metadata": {},
   "source": [
    "# Expression of transcripts in GTEx Long-read dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c4e5a7d",
   "metadata": {},
   "source": [
    "### Import 2023 reference and disease gene lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dfa4bd6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open original reference\n",
    "ref_2023 = pd.read_csv(\"../../references/Homo_sapiens.GRCh38.109.gtf\", header=None, delimiter=\"\\t\", low_memory=False, \n",
    "                       names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"], comment=\"#\")\n",
    "\n",
    "# filter out anything with ERCC (shouldn't be there, but just in case)\n",
    "ref_2023 = ref_2023.loc[~ref_2023[\"chr\"].str.startswith(\"ERCC-\")]\n",
    "\n",
    "## Parse through reference to get gene names and ids\n",
    "gene_ref_2023 = ref_2023.loc[ref_2023[\"type\"]==\"gene\"].copy()\n",
    "gene_ref_2023 = parse_df_columns(gene_ref_2023, is_ref=True)\n",
    "gene_info = gene_ref_2023[['gene_id', 'gene_biotype', 'gene_name']].copy()\n",
    "\n",
    "## Create 2023 transcript ref\n",
    "transcript_ref_2023 = ref_2023.loc[ref_2023[\"type\"]==\"transcript\"].copy()\n",
    "transcript_ref_2023 = parse_df_columns(transcript_ref_2023, is_ref=True, is_transcript=True)\n",
    "# get the transcript biotypes\n",
    "transcript_biotype = transcript_ref_2023[['transcript_id', 'transcript_biotype']].copy()\n",
    "\n",
    "## Import disease relevant genes\n",
    "disease_relevant_genes = pd.read_csv(\"../../references/medically_relevant_genes.tsv\", sep=\"\\t\")\n",
    "\n",
    "## Brain disease genes\n",
    "brain_disease_gene_ids = pd.read_csv(\"../../references/brain_disease_genes_only_IDs.tsv\", sep=\"\\t\")\n",
    "brain_disease_annotations = pd.read_csv(\"../../references/brain_disease_genes_with_disease.tsv\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "## Create disease relevant list including chromosome\n",
    "disease_relevant_genes_annotated = disease_relevant_genes.merge(gene_ref_2023[[\"gene_id\", \"gene_name\", \"chr\"]], \n",
    "                                                               how=\"inner\", on=[\"gene_id\", \"gene_name\"])\n",
    "\n",
    "## Create list of protein coding genes\n",
    "protein_coding_ref_2023 = gene_ref_2023.loc[gene_ref_2023[\"gene_biotype\"] == \"protein_coding\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "844d7fbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>n_anno_tx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    gene_biotype gene_name  n_anno_tx\n",
       "0  ENSG00000000003  protein_coding    TSPAN6          5\n",
       "1  ENSG00000000005  protein_coding      TNMD          2\n",
       "2  ENSG00000000419  protein_coding      DPM1         16\n",
       "3  ENSG00000000457  protein_coding     SCYL3          5\n",
       "4  ENSG00000000460  protein_coding  C1orf112          9"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get number of annotated isoforms per gene\n",
    "annotated_n_iso_per_gene = transcript_ref_2023.copy()[['transcript_id', 'gene_id', 'gene_biotype', 'gene_name']].groupby([\"gene_id\", 'gene_biotype', \"gene_name\"]).size().reset_index(name='n_anno_tx')\n",
    "annotated_n_iso_per_gene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80234ded",
   "metadata": {},
   "source": [
    "## Create reference for protein coding sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed68b97d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make reference only including protein coding transcripts with a defined CDS\n",
    "ref_2023_cds = ref_2023.loc[ref_2023[\"type\"]==\"CDS\"].copy()\n",
    "ref_2023_cds = parse_df_columns(ref_2023_cds.copy(), is_ref=True, is_transcript=True, is_prot=True)\n",
    "\n",
    "ref_2023_cds = ref_2023_cds.loc[ref_2023_cds[\"transcript_biotype\"] == \"protein_coding\"].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "559b2df8",
   "metadata": {},
   "source": [
    "#### Create unique identifier for protein id based on individual CDS entries for the protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "20492add",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create datagrame for unique protein id \n",
    "unique_protein_id = pd.DataFrame()\n",
    "\n",
    "## Create a reference with columns as strings instead of integers so they can be concatenated\n",
    "string_ref_2023_cds = ref_2023_cds[['protein_id', 'start', 'end', \"chr\", \"strand\"]].copy().astype(str)\n",
    "\n",
    "## Join the coordinates for all CDS based on protein id to create a unique identifier\n",
    "unique_protein_id = string_ref_2023_cds.groupby(['protein_id'])[[\"protein_id\", \"chr\", \"strand\",\n",
    "                                                                 \"start\", \"end\"]].transform(lambda x: '-'.join(x))\n",
    "\n",
    "## Add protein id column\n",
    "unique_protein_id[\"protein_id\"] = unique_protein_id[\"protein_id\"].str.split(\"-\", expand=True)[0]\n",
    "\n",
    "## Add unique identifier column\n",
    "unique_protein_id[\"unique_identifier\"] = \"Chromosome: \" + unique_protein_id[\"chr\"].str.split(\"-\", expand=True)[0] + \"\\n\" \\\n",
    "                                            + \"Strand: \" + unique_protein_id[\"strand\"].str.split(\"-\", expand=True)[0] + \"\\n\" \\\n",
    "                                            + \"Start coordinates: \" + unique_protein_id[\"start\"].copy() \\\n",
    "                                            + \"End coordinates:\" + unique_protein_id[\"end\"].copy()\n",
    "\n",
    "## Drop all columns besided protein id and unique identifier\n",
    "unique_protein_id = unique_protein_id[[\"protein_id\", \"unique_identifier\"]].copy()\n",
    "\n",
    "## Drop any duplicates\n",
    "unique_protein_id.drop_duplicates(inplace=True)\n",
    "\n",
    "## Get transcript_id column back by merging on index with ref_2023_cds\n",
    "unique_protein_id = unique_protein_id.merge(ref_2023_cds[\"transcript_id\"], left_index=True, right_index=True)\n",
    "\n",
    "## Merge back to ref_2023_cds on protein_id and transcript id to so we can add unique identifier\n",
    "ref_2023_cds = ref_2023_cds.merge(unique_protein_id, on=[\"protein_id\", \"transcript_id\"], how=\"inner\")\n",
    "\n",
    "## Make protein coding ID into unique identifier\n",
    "ref_2023_cds[\"protein_id\"] = ref_2023_cds[\"unique_identifier\"].copy()\n",
    "ref_2023_cds.drop(columns=\"unique_identifier\", inplace=True)\n",
    "\n",
    "## Create final CDS reference only containing non NA valued for protein id\n",
    "ref_cds = ref_2023_cds.loc[~ref_2023_cds[\"protein_id\"].isna()].copy()\n",
    "\n",
    "## Only keep transcript id and unique protein id in the ref_cds final reference\n",
    "ref_cds = ref_cds[[\"transcript_id\", \"protein_id\"]].copy().drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e412e2db",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Output references with transcripts_id and unique protein_id as a tsv to the processed data folder\n",
    "\n",
    "ref_cds.to_csv(\"../../tables/transcript_id_and_unique_protein_id_reference.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e96bcef0",
   "metadata": {},
   "source": [
    "## Load transcript counts matrix and CPM normalize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7796e004",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples we are not using due to replicates, low read depth, and poor PCA clustering\n",
    "filter_out_samples = [ \n",
    "    \"GTEX-Q2AG-0011-R11A-SM-2EBL2_rep2.FAK44637.bam\", \n",
    "    \"GTEX-Q2AG-0011-R11A-SM-2EBL2_rep.FAK49243.bam\", \n",
    "    \"GTEX-T5JC-0011-R10A-SM-2TT23.FAK91589.bam\", \n",
    "    \"GTEX-QEG5-0008-SM-3QHW2_exp.FAK30166.bam\", \n",
    "    \"GTEX-QV44-0008-SM-3QNG7_ctrl1.FAK55556.bam\",\n",
    "    \"GTEX-QV44-0008-SM-3QNG7_exp.FAK52124.bam\",\n",
    "    \"GTEX-RWS6-0008-SM-3QHWG_rep.FAK49207.bam\", \n",
    "    \"GTEX-S4Z8-0008-SM-2Y983_exp1.FAK55723.bam\",\n",
    "    \"GTEX-S4Z8-0008-SM-2Y983_exp2.FAK47416.bam\",\n",
    "    \"GTEX-S95S-0008-SM-3RQ8B_exp1.FAK55217.bam\",\n",
    "    \"GTEX-S95S-0008-SM-3RQ8B_exp2.FAK47088.bam\",\n",
    "    \"GTEX-WY7C-0008-SM-3NZB5_ctrl.FAK55679.bam\",\n",
    "    \"GTEX-1GN1W-0226-SM-7AGLJ_rep.FAK91654.bam\",\n",
    "    \"GTEX-WY7C-1126-SM-3GS2X_rep2.FAK49168.bam\",\n",
    "    \"GTEX-WY7C-1126-SM-3GS2X.FAK39149.bam\",\n",
    "    \"GTEX-Y5LM-0426-SM-3YX99.FAK52212.bam\",\n",
    "    \"GTEX-Y5LM-0426-SM-3YX99_rep2.FAK41279.bam\",\n",
    "    \"GTEX-14BMU-0526-SM-5CA2F.FAK44778.bam\",\n",
    "    \"GTEX-14BMU-0526-SM-5CA2F_rep.FAK93376.bam\",\n",
    "    \"GTEX-13QJ3-0726-SM-7LDHS.FAK49189.bam\",\n",
    "    \"GTEX-ZT9X-1826-SM-4V2KV_rep.FAK39773.bam\",\n",
    "    \"GTEX-ZT9X-1826-SM-4V2KV.FAK49260.bam\", \n",
    "    \"GTEX-WY7C-0726-SM-3GLGQ.FAK46872.bam\" ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0a417410",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load matrix\n",
    "df = pd.read_csv(\"../../data/raw/GRCh38_quant_mapq10_gtf_109_and_high-confidence_GTEx_DATA/bambu_quant/counts_transcript.txt\", sep=\"\\t\")\n",
    "# load unique counts matrix\n",
    "df_unique = pd.read_csv(\"../../data/raw/GRCh38_quant_mapq10_gtf_109_and_high-confidence_GTEx_DATA/bambu_quant/uniqueCounts_transcript.txt\", sep=\"\\t\")\n",
    "# load gtex metadata\n",
    "gtex_metadata = pd.read_csv(\"/pscratch/mteb223_uksr/new_RNA_isoform_expression_across_tissues/data/GTEx_v9_ONT_metadata.txt\", sep = \"\\t\")\n",
    "\n",
    "## Convert sample names to bam files\n",
    "columns_to_rename = df.columns[2:]\n",
    "new_col_names = {col: get_samp_id(col) for col in columns_to_rename}\n",
    "df.rename(columns=new_col_names, inplace=True)\n",
    "df = df[df.columns.difference(filter_out_samples)]\n",
    "df_unique.rename(columns=new_col_names, inplace=True)\n",
    "df_unique = df_unique[df_unique.columns.difference(filter_out_samples)]\n",
    "\n",
    "# get the list of samples so that we can separate out the columns below by tissue\n",
    "df_cols_dict = {}\n",
    "for tissue in gtex_metadata[\"tissue_site_detail\"].copy().unique():\n",
    "    if tissue in tissues_to_use:\n",
    "        df_cols_dict[tissue] = [el for el in gtex_metadata.loc[gtex_metadata['tissue_site_detail'] == tissue, 'bam_file'].values.tolist() if not 'direct' in el]\n",
    "        df_cols_dict[tissue] = [x for x in df_cols_dict[tissue] if x not in filter_out_samples]\n",
    "\n",
    "tx_rel_ab = pd.DataFrame(columns = ['transcript_id', 'gene_id', 'gene_name'])\n",
    "df_by_tissue = {}\n",
    "\n",
    "# for each tissue\n",
    "for key in df_cols_dict:\n",
    "    # grab the columns we will need for this tissue\n",
    "    columns = [\"TXNAME\", \"GENEID\"] + df_cols_dict[key]\n",
    "    tmp = df[columns].copy()\n",
    "    \n",
    "    ## Fix columns names\n",
    "    tmp = fix_column_names(tmp, is_gene=False)\n",
    "    \n",
    "    ## Add total counts column\n",
    "    tmp[\"total_counts\"] = tmp[tmp.filter(regex='count').columns].copy().sum(axis=1)\n",
    "    \n",
    "    ## CPM normalize counts matrix\n",
    "    tmp = calculate_cpm(tmp, is_gene=False)\n",
    "    tmp[\"median_CPM\"] = round(tmp[tmp.filter(regex='CPM').columns].copy().median(axis=1), 2)\n",
    "    tmp.drop(columns=tmp.filter(regex='count').columns, inplace=True)\n",
    "    \n",
    "    tmp_u = df_unique[columns].copy()\n",
    "    \n",
    "    ## Fix columns names for unique counts\n",
    "    tmp_u = fix_column_names(tmp_u, is_gene=False)\n",
    "    ## Add median unique counts column\n",
    "    tmp_u[\"median_unique_counts\"] = tmp_u[tmp_u.filter(regex='count').columns].copy().median(axis=1)\n",
    "    # for the unique dataframe, keep only these columns for merging\n",
    "    tmp_u = tmp_u[['transcript_id', 'gene_id', 'median_unique_counts']].drop_duplicates()\n",
    "\n",
    "    # merge the all counts and unique counts together\n",
    "    tmp = pd.merge(tmp, tmp_u, how='inner')\n",
    "    # calculate the median relative transcript abundance\n",
    "    tmp = relative_transcript_abundance(tmp.copy())\n",
    "    # add the tissue name\n",
    "    tmp['tissue_site_detail'] = key\n",
    "    # add on the biotype of the transcript and gene\n",
    "    tmp = tmp.merge(transcript_biotype)\n",
    "    tmp = tmp.merge(gene_info)\n",
    "    tmp = tmp.merge(annotated_n_iso_per_gene)\n",
    "    tmp_rel_ab = tmp[['transcript_id', 'gene_id', 'gene_name', 'median_relative_abundance']].copy()\n",
    "    tmp_rel_ab = tmp_rel_ab.rename(columns={\"median_relative_abundance\": key})\n",
    "    tx_rel_ab = tx_rel_ab.merge(tmp_rel_ab, how='outer')\n",
    "    # write table to file\n",
    "    tmp.to_csv(\"../../data/processed/GTEx/GTEX_\" + key + \"_cpm_transcript.tsv\", sep=\"\\t\", index=False)\n",
    "    # filter -> transcripts must be present with at least one unique count in at least half the samples\n",
    "    tmp = tmp.loc[tmp[\"median_unique_counts\"] >= 1].copy()\n",
    "    df_by_tissue[key] = tmp\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a2d2bc2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>GTEX-1211K-0826-SM-7LDFQ.FAK46515.bam_CPM</th>\n",
       "      <th>GTEX-14BMU-0526-SM-5CA2F_rep2.FAK49039.bam_CPM</th>\n",
       "      <th>GTEX-1I6K7-1226-SM-AAEQX.FAK44642.bam_CPM</th>\n",
       "      <th>GTEX-1KXAM-0426-SM-CYKMP.FAK44752.bam_CPM</th>\n",
       "      <th>GTEX-WYVS-0526-SM-3H5V7.FAK54827.bam_CPM</th>\n",
       "      <th>GTEX-ZT9X-0326-SM-4U9QG.FAK44894.bam_CPM</th>\n",
       "      <th>total_CPM</th>\n",
       "      <th>median_CPM</th>\n",
       "      <th>median_unique_counts</th>\n",
       "      <th>median_relative_abundance</th>\n",
       "      <th>median_gene_cpm</th>\n",
       "      <th>tissue_site_detail</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>n_anno_tx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.04</td>\n",
       "      <td>0.37</td>\n",
       "      <td>1.83</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2.0</td>\n",
       "      <td>26.768104</td>\n",
       "      <td>2.600</td>\n",
       "      <td>Lung</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000324913</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>89.16</td>\n",
       "      <td>68.25</td>\n",
       "      <td>72.72</td>\n",
       "      <td>88.75</td>\n",
       "      <td>83.30</td>\n",
       "      <td>86.92</td>\n",
       "      <td>80.68</td>\n",
       "      <td>83.30</td>\n",
       "      <td>275.0</td>\n",
       "      <td>96.518609</td>\n",
       "      <td>86.400</td>\n",
       "      <td>Lung</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000454841</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>0.45</td>\n",
       "      <td>1.47</td>\n",
       "      <td>2.22</td>\n",
       "      <td>3.91</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.79</td>\n",
       "      <td>1.48</td>\n",
       "      <td>1.47</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.483646</td>\n",
       "      <td>86.400</td>\n",
       "      <td>Lung</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000391616</td>\n",
       "      <td>ENSG00000240356</td>\n",
       "      <td>8.47</td>\n",
       "      <td>8.69</td>\n",
       "      <td>6.09</td>\n",
       "      <td>4.77</td>\n",
       "      <td>10.15</td>\n",
       "      <td>5.47</td>\n",
       "      <td>7.45</td>\n",
       "      <td>7.45</td>\n",
       "      <td>19.5</td>\n",
       "      <td>47.743446</td>\n",
       "      <td>13.315</td>\n",
       "      <td>Lung</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>RPL23AP7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000416673</td>\n",
       "      <td>ENSG00000240356</td>\n",
       "      <td>1.06</td>\n",
       "      <td>1.95</td>\n",
       "      <td>1.27</td>\n",
       "      <td>1.06</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.81</td>\n",
       "      <td>1.86</td>\n",
       "      <td>1.86</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.343428</td>\n",
       "      <td>13.315</td>\n",
       "      <td>Lung</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>RPL23AP7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id          gene_id  \\\n",
       "0  ENST00000488147  ENSG00000227232   \n",
       "1  ENST00000324913  ENSG00000125611   \n",
       "3  ENST00000454841  ENSG00000125611   \n",
       "6  ENST00000391616  ENSG00000240356   \n",
       "7  ENST00000416673  ENSG00000240356   \n",
       "\n",
       "   GTEX-1211K-0826-SM-7LDFQ.FAK46515.bam_CPM  \\\n",
       "0                                       0.00   \n",
       "1                                      89.16   \n",
       "3                                       0.45   \n",
       "6                                       8.47   \n",
       "7                                       1.06   \n",
       "\n",
       "   GTEX-14BMU-0526-SM-5CA2F_rep2.FAK49039.bam_CPM  \\\n",
       "0                                            0.42   \n",
       "1                                           68.25   \n",
       "3                                            1.47   \n",
       "6                                            8.69   \n",
       "7                                            1.95   \n",
       "\n",
       "   GTEX-1I6K7-1226-SM-AAEQX.FAK44642.bam_CPM  \\\n",
       "0                                       0.56   \n",
       "1                                      72.72   \n",
       "3                                       2.22   \n",
       "6                                       6.09   \n",
       "7                                       1.27   \n",
       "\n",
       "   GTEX-1KXAM-0426-SM-CYKMP.FAK44752.bam_CPM  \\\n",
       "0                                       1.04   \n",
       "1                                      88.75   \n",
       "3                                       3.91   \n",
       "6                                       4.77   \n",
       "7                                       1.06   \n",
       "\n",
       "   GTEX-WYVS-0526-SM-3H5V7.FAK54827.bam_CPM  \\\n",
       "0                                      0.37   \n",
       "1                                     83.30   \n",
       "3                                      0.19   \n",
       "6                                     10.15   \n",
       "7                                      2.42   \n",
       "\n",
       "   GTEX-ZT9X-0326-SM-4U9QG.FAK44894.bam_CPM  total_CPM  median_CPM  \\\n",
       "0                                      1.83       0.72        0.56   \n",
       "1                                     86.92      80.68       83.30   \n",
       "3                                      0.79       1.48        1.47   \n",
       "6                                      5.47       7.45        7.45   \n",
       "7                                      2.81       1.86        1.86   \n",
       "\n",
       "   median_unique_counts  median_relative_abundance  median_gene_cpm  \\\n",
       "0                   2.0                  26.768104            2.600   \n",
       "1                 275.0                  96.518609           86.400   \n",
       "3                   5.0                   1.483646           86.400   \n",
       "6                  19.5                  47.743446           13.315   \n",
       "7                   4.0                  10.343428           13.315   \n",
       "\n",
       "  tissue_site_detail       transcript_biotype  \\\n",
       "0               Lung   unprocessed_pseudogene   \n",
       "1               Lung           protein_coding   \n",
       "3               Lung  nonsense_mediated_decay   \n",
       "6               Lung     processed_transcript   \n",
       "7               Lung     processed_transcript   \n",
       "\n",
       "                       gene_biotype gene_name  n_anno_tx  \n",
       "0            unprocessed_pseudogene    WASH7P          1  \n",
       "1                    protein_coding    CHCHD5          5  \n",
       "3                    protein_coding    CHCHD5          5  \n",
       "6  transcribed_processed_pseudogene  RPL23AP7          3  \n",
       "7  transcribed_processed_pseudogene  RPL23AP7          3  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_by_tissue['Lung'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3a4a5d9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Brain - Frontal Cortex (BA9)\n",
      "['GTEX-1192X-0011-R10a-SM-4RXXZ.FAK49046.bam', 'GTEX-13X6J-0011-R10b-SM-5CEKT.FAK44896.bam', 'GTEX-14BIL-0011-R10a-SM-5EQV4.FAK49209.bam', 'GTEX-15DCD-0011-R10b-SM-5S51M.FAK42101.bam', 'GTEX-QDT8-0011-R10A-SM-2FKJB.FAK49182.bam']\n",
      "5\n",
      "Brain - Cerebellar Hemisphere\n",
      "['GTEX-11H98-0011-R11b-SM-4SFLZ.FAK46829.bam', 'GTEX-13VXU-0011-R11b-SM-5BFQZ.FAK44611.bam', 'GTEX-17F97-0011-R11b-SM-63KY2.FAK41775.bam', 'GTEX-1H3NZ-0011-R11b-SM-AUNOV.FAK49024.bam', 'GTEX-Q2AG-0011-R11A-SM-2EBL2.FAK42265.bam', 'GTEX-T5JC-0011-R11A-SM-2TT24.FAK54838.bam']\n",
      "6\n",
      "Brain - Putamen (basal ganglia)\n",
      "['GTEX-11TTK-0011-R7b-SM-4TVFS.FAK39197.bam', 'GTEX-1313W-0011-R7b-SM-4ZL3U.FAK44754.bam', 'GTEX-13RTJ-0011-R7b-SM-5CTCB.FAK49257.bam', 'GTEX-14C5O-0011-R7b-SM-5GUPO.FAK54887.bam', 'GTEX-15ER7-0011-R7a-SM-5QYP2.FAK44704.bam', 'GTEX-T5JC-0011-R7A-SM-2TT1Z.FAK42170.bam']\n",
      "6\n",
      "Lung\n",
      "['GTEX-1211K-0826-SM-7LDFQ.FAK46515.bam', 'GTEX-14BMU-0526-SM-5CA2F_rep2.FAK49039.bam', 'GTEX-1I6K7-1226-SM-AAEQX.FAK44642.bam', 'GTEX-1KXAM-0426-SM-CYKMP.FAK44752.bam', 'GTEX-WYVS-0526-SM-3H5V7.FAK54827.bam', 'GTEX-ZT9X-0326-SM-4U9QG.FAK44894.bam']\n",
      "6\n",
      "Heart - Left Ventricle\n",
      "['GTEX-13QBU-0426-SM-5A4VT.FAK49161.bam', 'GTEX-15RIE-1726-SM-5V7WB.FAK41916.bam', 'GTEX-1H11D-1226-SM-9P6XJ.FAK44759.bam', 'GTEX-1I6K7-0626-SM-AAEQW.FAK49296.bam', 'GTEX-1IDJU-0726-SM-BXZK2.FAK49131.bam', 'GTEX-OHPL-0326-SM-DPSB8.FAK54495.bam', 'GTEX-ZVZP-0226-SM-4VEIO.FAK41904.bam']\n",
      "7\n",
      "Muscle - Skeletal\n",
      "['GTEX-13QJ3-0726-SM-7LDHS_rep.FAK93371.bam', 'GTEX-17MFQ-1926-SM-6659L.FAK44757.bam', 'GTEX-1C64N-0326-SM-6ZHT1.FAK53752.bam', 'GTEX-1KXAM-2426-SM-CYKMT.FAK39676.bam', 'GTEX-1LVA9-0326-SM-CYRXA.FAK39191.bam', 'GTEX-ZT9X-1826-SM-4V2KV_rep2.FAK44613.bam']\n",
      "6\n",
      "Heart - Atrial Appendage\n",
      "['GTEX-14XAO-0926-SM-5YQWC.FAK46869.bam', 'GTEX-1GN1W-0226-SM-7AGLJ.FAK44655.bam', 'GTEX-1HBPH-0226-SM-7LLUW.FAK41473.bam', 'GTEX-1IDJD-0226-SM-AML89.FAK49187.bam', 'GTEX-1IDJF-0826-SM-ABADH.FAK39145.bam', 'GTEX-WY7C-1126-SM-3GS2X_rep.FAK49218.bam']\n",
      "6\n",
      "Cells - Cultured fibroblasts\n",
      "['GTEX-OIZI-0008-SM-2FR3P.FAK41829.bam', 'GTEX-OXRL-0008-SM-2FR3T.FAK39703.bam', 'GTEX-PSDG-0008-SM-3NZBI.FAK41823.bam', 'GTEX-QEG5-0008-SM-3QHW2_ctrl.FAK55175.bam', 'GTEX-QV44-0008-SM-3QNG7_ctrl2.FAK49435.bam', 'GTEX-R55E-0008-SM-3QNG3.FAK46687.bam', 'GTEX-RWS6-0008-SM-3QHWG.FAK46867.bam', 'GTEX-S4Z8-0008-SM-2Y983_ctrl.FAK54991.bam', 'GTEX-S95S-0008-SM-3RQ8B_ctrl.FAK55194.bam', 'GTEX-WFG7-0008-SM-3ML7A.FAK41398.bam', 'GTEX-WY7C-0008-SM-3NZB5_ctrl.FAK55628.bam']\n",
      "11\n",
      "Liver\n",
      "['GTEX-R53T-0326-SM-2K8S4.FAK44579.bam', 'GTEX-UTHO-2426-SM-38ZXF.FAK46748.bam', 'GTEX-Y5LM-0426-SM-3YX99_rep.FAK49269.bam', 'GTEX-ZF29-2026-SM-4TVXH.FAK44621.bam', 'GTEX-ZPU1-0826-SM-4UJSC.FAK41797.bam']\n",
      "5\n",
      "58\n"
     ]
    }
   ],
   "source": [
    "# verify the number of samples being used from each tissue\n",
    "samps_we_use = []\n",
    "total_samples = 0\n",
    "for tiss in df_cols_dict:\n",
    "    # print the tissue, the sample ids, and the number of samples\n",
    "    print(tiss)\n",
    "    print(df_cols_dict[tiss])\n",
    "    samps_we_use = samps_we_use + df_cols_dict[tiss]\n",
    "    print(len(df_cols_dict[tiss]))\n",
    "    total_samples = total_samples + len(df_cols_dict[tiss])\n",
    "print(total_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "418dc0ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Brain - Frontal Cortex (BA9)</th>\n",
       "      <th>Brain - Cerebellar Hemisphere</th>\n",
       "      <th>Brain - Putamen (basal ganglia)</th>\n",
       "      <th>Lung</th>\n",
       "      <th>Heart - Left Ventricle</th>\n",
       "      <th>Muscle - Skeletal</th>\n",
       "      <th>Heart - Atrial Appendage</th>\n",
       "      <th>Cells - Cultured fibroblasts</th>\n",
       "      <th>Liver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>22.099448</td>\n",
       "      <td>13.800503</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26.768104</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>33.072917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000324913</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>77.539698</td>\n",
       "      <td>74.569664</td>\n",
       "      <td>81.238979</td>\n",
       "      <td>96.518609</td>\n",
       "      <td>84.485838</td>\n",
       "      <td>74.173987</td>\n",
       "      <td>85.993486</td>\n",
       "      <td>97.209181</td>\n",
       "      <td>98.010974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000409719</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.578364</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000454841</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>0.612302</td>\n",
       "      <td>2.023780</td>\n",
       "      <td>0.940918</td>\n",
       "      <td>1.483646</td>\n",
       "      <td>0.524374</td>\n",
       "      <td>0.574190</td>\n",
       "      <td>1.327411</td>\n",
       "      <td>1.531116</td>\n",
       "      <td>0.800183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000469030</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.260879</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id          gene_id gene_name  Brain - Frontal Cortex (BA9)  \\\n",
       "0  ENST00000488147  ENSG00000227232    WASH7P                     22.099448   \n",
       "1  ENST00000324913  ENSG00000125611    CHCHD5                     77.539698   \n",
       "2  ENST00000409719  ENSG00000125611    CHCHD5                      0.000000   \n",
       "3  ENST00000454841  ENSG00000125611    CHCHD5                      0.612302   \n",
       "4  ENST00000469030  ENSG00000125611    CHCHD5                      0.000000   \n",
       "\n",
       "   Brain - Cerebellar Hemisphere  Brain - Putamen (basal ganglia)       Lung  \\\n",
       "0                      13.800503                         0.000000  26.768104   \n",
       "1                      74.569664                        81.238979  96.518609   \n",
       "2                       0.578364                         0.000000   0.000000   \n",
       "3                       2.023780                         0.940918   1.483646   \n",
       "4                       0.000000                         0.000000   0.000000   \n",
       "\n",
       "   Heart - Left Ventricle  Muscle - Skeletal  Heart - Atrial Appendage  \\\n",
       "0                0.000000          50.000000                 33.072917   \n",
       "1               84.485838          74.173987                 85.993486   \n",
       "2                0.000000           0.000000                  0.000000   \n",
       "3                0.524374           0.574190                  1.327411   \n",
       "4                0.000000           0.000000                  0.000000   \n",
       "\n",
       "   Cells - Cultured fibroblasts      Liver  \n",
       "0                      0.000000  33.333333  \n",
       "1                     97.209181  98.010974  \n",
       "2                      0.000000   0.000000  \n",
       "3                      1.531116   0.800183  \n",
       "4                      0.260879   0.000000  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print the relative abundance to file\n",
    "tx_rel_ab.to_csv('../../tables/GTEx_expression/GTEx_tx_relative_abundance.tsv', sep='\\t', index=False)\n",
    "tx_rel_ab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cc5a70fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENEID</th>\n",
       "      <th>TXNAME</th>\n",
       "      <th>GTEX-1192X-0011-R10a-SM-4RXXZ.FAK49046.bam</th>\n",
       "      <th>GTEX-13X6J-0011-R10b-SM-5CEKT.FAK44896.bam</th>\n",
       "      <th>GTEX-14BIL-0011-R10a-SM-5EQV4.FAK49209.bam</th>\n",
       "      <th>GTEX-15DCD-0011-R10b-SM-5S51M.FAK42101.bam</th>\n",
       "      <th>GTEX-QDT8-0011-R10A-SM-2FKJB.FAK49182.bam</th>\n",
       "      <th>GTEX-11H98-0011-R11b-SM-4SFLZ.FAK46829.bam</th>\n",
       "      <th>GTEX-13VXU-0011-R11b-SM-5BFQZ.FAK44611.bam</th>\n",
       "      <th>GTEX-17F97-0011-R11b-SM-63KY2.FAK41775.bam</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-WY7C-0008-SM-3NZB5_ctrl.FAK55628.bam</th>\n",
       "      <th>GTEX-R53T-0326-SM-2K8S4.FAK44579.bam</th>\n",
       "      <th>GTEX-UTHO-2426-SM-38ZXF.FAK46748.bam</th>\n",
       "      <th>GTEX-Y5LM-0426-SM-3YX99_rep.FAK49269.bam</th>\n",
       "      <th>GTEX-ZF29-2026-SM-4TVXH.FAK44621.bam</th>\n",
       "      <th>GTEX-ZPU1-0826-SM-4UJSC.FAK41797.bam</th>\n",
       "      <th>total_unique_counts</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>BambuTx1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>21</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>193</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>BambuTx100</td>\n",
       "      <td>52</td>\n",
       "      <td>8</td>\n",
       "      <td>27</td>\n",
       "      <td>20</td>\n",
       "      <td>35</td>\n",
       "      <td>36</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>1</td>\n",
       "      <td>2901</td>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>BambuTx95</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>BambuTx97</td>\n",
       "      <td>27</td>\n",
       "      <td>31</td>\n",
       "      <td>66</td>\n",
       "      <td>8</td>\n",
       "      <td>136</td>\n",
       "      <td>38</td>\n",
       "      <td>308</td>\n",
       "      <td>48</td>\n",
       "      <td>...</td>\n",
       "      <td>26</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>5557</td>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GENEID           TXNAME  \\\n",
       "0  ENSG00000227232         BambuTx1   \n",
       "1  ENSG00000227232  ENST00000488147   \n",
       "2  ENSG00000215861       BambuTx100   \n",
       "3  ENSG00000215861        BambuTx95   \n",
       "4  ENSG00000215861        BambuTx97   \n",
       "\n",
       "   GTEX-1192X-0011-R10a-SM-4RXXZ.FAK49046.bam  \\\n",
       "0                                           4   \n",
       "1                                           1   \n",
       "2                                          52   \n",
       "3                                           0   \n",
       "4                                          27   \n",
       "\n",
       "   GTEX-13X6J-0011-R10b-SM-5CEKT.FAK44896.bam  \\\n",
       "0                                           1   \n",
       "1                                           0   \n",
       "2                                           8   \n",
       "3                                           0   \n",
       "4                                          31   \n",
       "\n",
       "   GTEX-14BIL-0011-R10a-SM-5EQV4.FAK49209.bam  \\\n",
       "0                                          10   \n",
       "1                                           8   \n",
       "2                                          27   \n",
       "3                                           0   \n",
       "4                                          66   \n",
       "\n",
       "   GTEX-15DCD-0011-R10b-SM-5S51M.FAK42101.bam  \\\n",
       "0                                           7   \n",
       "1                                           2   \n",
       "2                                          20   \n",
       "3                                           0   \n",
       "4                                           8   \n",
       "\n",
       "   GTEX-QDT8-0011-R10A-SM-2FKJB.FAK49182.bam  \\\n",
       "0                                         21   \n",
       "1                                          7   \n",
       "2                                         35   \n",
       "3                                          0   \n",
       "4                                        136   \n",
       "\n",
       "   GTEX-11H98-0011-R11b-SM-4SFLZ.FAK46829.bam  \\\n",
       "0                                           5   \n",
       "1                                           1   \n",
       "2                                          36   \n",
       "3                                           0   \n",
       "4                                          38   \n",
       "\n",
       "   GTEX-13VXU-0011-R11b-SM-5BFQZ.FAK44611.bam  \\\n",
       "0                                           1   \n",
       "1                                           7   \n",
       "2                                           5   \n",
       "3                                           0   \n",
       "4                                         308   \n",
       "\n",
       "   GTEX-17F97-0011-R11b-SM-63KY2.FAK41775.bam  ...  \\\n",
       "0                                           3  ...   \n",
       "1                                           6  ...   \n",
       "2                                          13  ...   \n",
       "3                                           0  ...   \n",
       "4                                          48  ...   \n",
       "\n",
       "   GTEX-WY7C-0008-SM-3NZB5_ctrl.FAK55628.bam  \\\n",
       "0                                          0   \n",
       "1                                          0   \n",
       "2                                         33   \n",
       "3                                          0   \n",
       "4                                         26   \n",
       "\n",
       "   GTEX-R53T-0326-SM-2K8S4.FAK44579.bam  GTEX-UTHO-2426-SM-38ZXF.FAK46748.bam  \\\n",
       "0                                     2                                     0   \n",
       "1                                     0                                     0   \n",
       "2                                     2                                     0   \n",
       "3                                     0                                     0   \n",
       "4                                     4                                     1   \n",
       "\n",
       "   GTEX-Y5LM-0426-SM-3YX99_rep.FAK49269.bam  \\\n",
       "0                                         1   \n",
       "1                                         1   \n",
       "2                                         0   \n",
       "3                                         0   \n",
       "4                                        13   \n",
       "\n",
       "   GTEX-ZF29-2026-SM-4TVXH.FAK44621.bam  GTEX-ZPU1-0826-SM-4UJSC.FAK41797.bam  \\\n",
       "0                                     2                                     0   \n",
       "1                                     1                                     1   \n",
       "2                                    15                                     1   \n",
       "3                                     0                                     0   \n",
       "4                                     3                                     1   \n",
       "\n",
       "   total_unique_counts          gene_id                        gene_biotype  \\\n",
       "0                  193  ENSG00000227232              unprocessed_pseudogene   \n",
       "1                   95  ENSG00000227232              unprocessed_pseudogene   \n",
       "2                 2901  ENSG00000215861  transcribed_unprocessed_pseudogene   \n",
       "3                    6  ENSG00000215861  transcribed_unprocessed_pseudogene   \n",
       "4                 5557  ENSG00000215861  transcribed_unprocessed_pseudogene   \n",
       "\n",
       "   gene_name  \n",
       "0     WASH7P  \n",
       "1     WASH7P  \n",
       "2        NaN  \n",
       "3        NaN  \n",
       "4        NaN  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# calculate the total unique counts across all used samples for each transcript\n",
    "filter_col = [\"GENEID\", \"TXNAME\"] + samps_we_use\n",
    "unique_counts = df_unique[filter_col].copy()\n",
    "unique_counts['total_unique_counts'] = unique_counts.select_dtypes(include=[np.number]).sum(axis=1)\n",
    "unique_counts = pd.merge(unique_counts, gene_info, left_on='GENEID', right_on='gene_id')\n",
    "unique_counts.to_csv('../../tables/GTEx_expression/GTEx_total_unique_counts_all_samples_by_tx.tsv', sep='\\t', index=False)\n",
    "unique_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a840b93f",
   "metadata": {},
   "source": [
    "## Look at the number of isoforms expressed in tissue vs gene CPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2e799eb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Brain - Frontal Cortex (BA9)\n",
      "Brain - Cerebellar Hemisphere\n",
      "Brain - Putamen (basal ganglia)\n",
      "Lung\n",
      "Heart - Left Ventricle\n",
      "Muscle - Skeletal\n",
      "Heart - Atrial Appendage\n",
      "Cells - Cultured fibroblasts\n",
      "Liver\n"
     ]
    }
   ],
   "source": [
    "# look at the number of isoforms expressed in each tissue\n",
    "cpm_vs_n_iso_diff_annotated = pd.DataFrame(columns=['gene_id', 'gene_biotype', 'median_gene_cpm', 'n_tx', 'n_anno_tx', 'n_tx_diff', 'tissue'])\n",
    "# for each tissue\n",
    "for tissue in df_by_tissue:\n",
    "    print(tissue)\n",
    "    tmp = df_by_tissue[tissue].copy()\n",
    "    tmp = tmp.copy()[['transcript_id', 'gene_id', 'median_gene_cpm', 'gene_biotype', 'n_anno_tx']]\n",
    "    # get the number of expressed isoforms\n",
    "    n_iso_tmp = tmp.copy().groupby([\"gene_id\", 'gene_biotype', \"median_gene_cpm\", 'n_anno_tx']).size().reset_index(name='n_tx')\n",
    "    n_iso_tmp['tissue'] = tissue\n",
    "    n_iso_tmp['n_tx_diff'] = n_iso_tmp['n_anno_tx'] - n_iso_tmp['n_tx']\n",
    "    cpm_vs_n_iso_diff_annotated = pd.concat([cpm_vs_n_iso_diff_annotated, n_iso_tmp])\n",
    "\n",
    "cpm_vs_n_iso_diff_annotated.to_csv('../../tables/GTEx_expression/GTEx_n_iso_diff_vs_gene_CPM.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e817c6b",
   "metadata": {},
   "source": [
    "## Make tables looking at tissue expression of isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c7c6156e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.01\n",
      "1.01\n",
      "5.01\n",
      "10.01\n",
      "Brain - Frontal Cortex (BA9)\n",
      "           gene_id    gene_biotype gene_name  Brain - Frontal Cortex (BA9)\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                             1\n",
      "1  ENSG00000000005  protein_coding      TNMD                             1\n",
      "2  ENSG00000000419  protein_coding      DPM1                             5\n",
      "3  ENSG00000000457  protein_coding     SCYL3                             1\n",
      "4  ENSG00000001036  protein_coding     FUCA2                             1\n",
      "Brain - Cerebellar Hemisphere\n",
      "           gene_id    gene_biotype gene_name  Brain - Cerebellar Hemisphere\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                              1\n",
      "1  ENSG00000000419  protein_coding      DPM1                              4\n",
      "2  ENSG00000000457  protein_coding     SCYL3                              1\n",
      "3  ENSG00000001036  protein_coding     FUCA2                              1\n",
      "4  ENSG00000001084  protein_coding      GCLC                              2\n",
      "Brain - Putamen (basal ganglia)\n",
      "           gene_id    gene_biotype gene_name  Brain - Putamen (basal ganglia)\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                                1\n",
      "1  ENSG00000000419  protein_coding      DPM1                                4\n",
      "2  ENSG00000000457  protein_coding     SCYL3                                1\n",
      "3  ENSG00000000971  protein_coding       CFH                                1\n",
      "4  ENSG00000001036  protein_coding     FUCA2                                1\n",
      "Lung\n",
      "           gene_id    gene_biotype gene_name  Lung\n",
      "0  ENSG00000000003  protein_coding    TSPAN6     1\n",
      "1  ENSG00000000419  protein_coding      DPM1     5\n",
      "2  ENSG00000000457  protein_coding     SCYL3     1\n",
      "3  ENSG00000000938  protein_coding       FGR     2\n",
      "4  ENSG00000000971  protein_coding       CFH     2\n",
      "Heart - Left Ventricle\n",
      "           gene_id    gene_biotype gene_name  Heart - Left Ventricle\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                       1\n",
      "1  ENSG00000000419  protein_coding      DPM1                       4\n",
      "2  ENSG00000000457  protein_coding     SCYL3                       1\n",
      "3  ENSG00000000971  protein_coding       CFH                       1\n",
      "4  ENSG00000001036  protein_coding     FUCA2                       1\n",
      "Muscle - Skeletal\n",
      "           gene_id    gene_biotype gene_name  Muscle - Skeletal\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                  1\n",
      "1  ENSG00000000005  protein_coding      TNMD                  1\n",
      "2  ENSG00000000419  protein_coding      DPM1                  5\n",
      "3  ENSG00000000457  protein_coding     SCYL3                  1\n",
      "4  ENSG00000000971  protein_coding       CFH                  1\n",
      "Heart - Atrial Appendage\n",
      "           gene_id    gene_biotype gene_name  Heart - Atrial Appendage\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                         1\n",
      "1  ENSG00000000419  protein_coding      DPM1                         4\n",
      "2  ENSG00000000457  protein_coding     SCYL3                         1\n",
      "3  ENSG00000000971  protein_coding       CFH                         4\n",
      "4  ENSG00000001036  protein_coding     FUCA2                         1\n",
      "Cells - Cultured fibroblasts\n",
      "           gene_id    gene_biotype gene_name  Cells - Cultured fibroblasts\n",
      "0  ENSG00000000003  protein_coding    TSPAN6                             1\n",
      "1  ENSG00000000419  protein_coding      DPM1                             5\n",
      "2  ENSG00000000457  protein_coding     SCYL3                             1\n",
      "3  ENSG00000000971  protein_coding       CFH                             2\n",
      "4  ENSG00000001036  protein_coding     FUCA2                             1\n",
      "Liver\n",
      "           gene_id    gene_biotype gene_name  Liver\n",
      "0  ENSG00000000003  protein_coding    TSPAN6      1\n",
      "1  ENSG00000000419  protein_coding      DPM1      3\n",
      "2  ENSG00000000457  protein_coding     SCYL3      1\n",
      "3  ENSG00000000971  protein_coding       CFH      9\n",
      "4  ENSG00000001036  protein_coding     FUCA2      1\n"
     ]
    }
   ],
   "source": [
    "# initialize variables\n",
    "# list of isoforms and protein_coding isoforms\n",
    "df_transcripts = {}\n",
    "df_ccds = {}\n",
    "\n",
    "# create list of thresholds\n",
    "list_cpm_thresh = [x/100 for x in range(0,1002)]\n",
    "\n",
    "print(list_cpm_thresh[1])\n",
    "print(list_cpm_thresh[101])\n",
    "print(list_cpm_thresh[501])\n",
    "print(list_cpm_thresh[1001])\n",
    "\n",
    "## Create lists with numbers of transcripts expressed across CPM thresholds for different categories\n",
    "list_2023_all_transcript_median = {}\n",
    "list_2023_all_gene_median = {}\n",
    "list_2023_cds_transcript_median = {}\n",
    "list_2023_med_relevant_transcript_median = {}\n",
    "list_2023_med_relevant_cds_transcript_median = {}\n",
    "list_2023_brain_relevant_transcript_median = {}\n",
    "list_2023_brain_relevant_cds_transcript_median = {}\n",
    "\n",
    "# look at the number of isoforms expressed in each tissue\n",
    "isoforms_in_tissue = pd.DataFrame(columns=['transcript_id', 'gene_id'])\n",
    "\n",
    "values_at_cpm_threshold = pd.DataFrame(columns=['cpm_threshold', 'tissue', \n",
    "                                                'all_transcript_median', \n",
    "                                                'all_gene_median',\n",
    "                                                'new_transcript_median', \n",
    "                                                'med_relevant_transcript_median', \n",
    "                                               'med_relevant_cds_transcript_median',\n",
    "                                               'brain_relevant_transcript_median',\n",
    "                                               'brain_relevant_cds_transcript_median'])\n",
    "\n",
    "# create dataframes for number of isoforms per gene (or protein-coding gene) by tissue\n",
    "n_tx_per_gene_by_tissue = pd.DataFrame(columns=['gene_id', 'gene_biotype'])\n",
    "n_tx_per_pc_gene_by_tissue = pd.DataFrame(columns=['gene_id', 'gene_biotype'])\n",
    "\n",
    "# create sets of isoforms at CPM gt 1\n",
    "sets_of_isoforms_at_cpm_gt_1 = {}\n",
    "sets_of_pc_isoforms_at_cpm_gt_1 = {}\n",
    "\n",
    "for tissue in df_by_tissue:\n",
    "    print(tissue)\n",
    "    \n",
    "    tmp = df_by_tissue[tissue].copy()\n",
    "    df_transcripts[tissue] = tmp.copy()\n",
    "    df_ccds[tissue] = tmp.loc[tmp[\"transcript_id\"].isin(ref_cds[\"transcript_id\"])].copy()\n",
    "    \n",
    "    # create empty list for the tissue\n",
    "    # all isoforms\n",
    "    list_2023_all_transcript_median[tissue] = []\n",
    "    # all genes\n",
    "    list_2023_all_gene_median[tissue] = []\n",
    "    # all protein-coding isoforms\n",
    "    list_2023_cds_transcript_median[tissue] = []\n",
    "    # all isoforms from medically relevant genes\n",
    "    list_2023_med_relevant_transcript_median[tissue] = []\n",
    "    # all isoforms from medically relevant protein-coding genes\n",
    "    list_2023_med_relevant_cds_transcript_median[tissue] = []\n",
    "    # all isoforms from brain disease relevant genes\n",
    "    list_2023_brain_relevant_transcript_median[tissue] = []\n",
    "    # all isoforms from brain disease relevant protein-coding genes\n",
    "    list_2023_brain_relevant_cds_transcript_median[tissue] = []\n",
    "    \n",
    "    # for a range of CPM thresholds\n",
    "    for i in range(0, 1002):\n",
    "\n",
    "        cpm_thresh = i/100\n",
    "        \n",
    "        # get the isoforms that fit each category\n",
    "        median_2023 = tmp.loc[tmp[\"median_CPM\"] >= cpm_thresh].copy()\n",
    "        cds_median_2023 = df_ccds[tissue].loc[df_ccds[tissue][\"median_CPM\"] >= cpm_thresh].copy()\n",
    "        med_relevant_median_2023 = median_2023.loc[median_2023[\"gene_id\"].isin(disease_relevant_genes[\"gene_id\"])].copy()\n",
    "        med_relevant_cds_median_2023 = med_relevant_median_2023.loc[med_relevant_median_2023[\"transcript_id\"].isin(cds_median_2023[\"transcript_id\"])].copy()\n",
    "        brain_relevant_median_2023 = median_2023.loc[median_2023[\"gene_id\"].isin(brain_disease_gene_ids[\"gene_id\"])].copy()\n",
    "        brain_relevant_cds_median_2023 = brain_relevant_median_2023.loc[brain_relevant_median_2023[\"transcript_id\"].isin(cds_median_2023[\"transcript_id\"])].copy()\n",
    "             \n",
    "        # at CPM > 1 (in this case, because we round values to 2 decimals, CPM == 1.01 is equivalent to CPM > 1)\n",
    "        if cpm_thresh == 1.01:\n",
    "            n_tx_per_gene_tmp = median_2023.copy()\n",
    "            # get the set of all isoforms expressed in this tissue above our thresholds\n",
    "            sets_of_isoforms_at_cpm_gt_1[tissue] = set(n_tx_per_gene_tmp.copy()['transcript_id'])\n",
    "            # get the median CPM for each isoform and then merge into larger dataframe\n",
    "            tmp_isoforms_in_tissue = n_tx_per_gene_tmp.copy()[['transcript_id', 'gene_id', 'gene_name', 'gene_biotype', 'median_CPM']]\n",
    "            tmp_isoforms_in_tissue = tmp_isoforms_in_tissue.rename(columns={'median_CPM': tissue})\n",
    "            isoforms_in_tissue = isoforms_in_tissue.merge(tmp_isoforms_in_tissue, how='outer')\n",
    "            # calculate the number of isoforms per gene in the tissue\n",
    "            n_tx_per_gene_tmp = n_tx_per_gene_tmp.groupby(['gene_id', 'gene_biotype', 'gene_name']).size().reset_index(name=\"n_tx\")\n",
    "            n_tx_per_gene_tmp = n_tx_per_gene_tmp.rename(columns={\"n_tx\": tissue})\n",
    "            n_tx_per_gene_by_tissue = n_tx_per_gene_by_tissue.merge(n_tx_per_gene_tmp, how='outer')\n",
    "            print(n_tx_per_gene_tmp.head())\n",
    "\n",
    "            # looking at protein-coding isoforms\n",
    "            n_tx_per_pc_gene_tmp = cds_median_2023.copy()\n",
    "            # get the set of all protein-coding isoforms expressed in this tissue above our thresholds\n",
    "            sets_of_pc_isoforms_at_cpm_gt_1[tissue] = set(n_tx_per_pc_gene_tmp.copy()['transcript_id'])\n",
    "            # get the number of protein-coding isoforms per gene in the tissue\n",
    "            n_tx_per_pc_gene_tmp = n_tx_per_pc_gene_tmp.groupby(['gene_id', 'gene_biotype', 'gene_name']).size().reset_index(name=\"n_tx\")\n",
    "            n_tx_per_pc_gene_tmp = n_tx_per_pc_gene_tmp.rename(columns={\"n_tx\": tissue})\n",
    "            n_tx_per_pc_gene_by_tissue = n_tx_per_pc_gene_by_tissue.merge(n_tx_per_pc_gene_tmp, how='outer')\n",
    "\n",
    "        # add just the total number in each category to the list\n",
    "        list_2023_all_transcript_median[tissue].append(median_2023.shape[0])\n",
    "        list_2023_all_gene_median[tissue].append(median_2023['gene_id'].nunique())\n",
    "        list_2023_cds_transcript_median[tissue].append(cds_median_2023.shape[0])\n",
    "        list_2023_med_relevant_transcript_median[tissue].append(med_relevant_median_2023.shape[0])\n",
    "        list_2023_med_relevant_cds_transcript_median[tissue].append(med_relevant_cds_median_2023.shape[0])\n",
    "        list_2023_brain_relevant_transcript_median[tissue].append(brain_relevant_median_2023.shape[0])\n",
    "        list_2023_brain_relevant_cds_transcript_median[tissue].append(brain_relevant_cds_median_2023.shape[0])\n",
    "    \n",
    "    # concat values to dataframe holding all the numbers for the thresholds for all tissues\n",
    "    values_at_cpm_threshold = pd.concat([values_at_cpm_threshold, pd.DataFrame({\n",
    "        'cpm_threshold': list_cpm_thresh,\n",
    "        'tissue': tissue,\n",
    "        'all_transcript_median': list_2023_all_transcript_median[tissue],\n",
    "        'all_gene_median': list_2023_all_gene_median[tissue],\n",
    "        'cds_transcript_median': list_2023_cds_transcript_median[tissue],\n",
    "        'med_relevant_transcript_median': list_2023_med_relevant_transcript_median[tissue],\n",
    "        'med_relevant_cds_transcript_median': list_2023_med_relevant_cds_transcript_median[tissue],\n",
    "        'brain_relevant_transcript_median': list_2023_brain_relevant_transcript_median[tissue],\n",
    "        'brain_relevant_cds_transcript_median': list_2023_brain_relevant_cds_transcript_median[tissue]\n",
    "    })])\n",
    "\n",
    "# write dataframe to file\n",
    "values_at_cpm_threshold.to_csv(\"../../tables/GTEx_expression/gtex_values_at_cpm_thresholds.tsv\", sep = '\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "65524153",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Brain - Frontal Cortex (BA9)</th>\n",
       "      <th>Brain - Cerebellar Hemisphere</th>\n",
       "      <th>Brain - Putamen (basal ganglia)</th>\n",
       "      <th>Lung</th>\n",
       "      <th>Heart - Left Ventricle</th>\n",
       "      <th>Muscle - Skeletal</th>\n",
       "      <th>Heart - Atrial Appendage</th>\n",
       "      <th>Cells - Cultured fibroblasts</th>\n",
       "      <th>Liver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000324913</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>CHCHD5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>43.81</td>\n",
       "      <td>19.96</td>\n",
       "      <td>46.36</td>\n",
       "      <td>83.30</td>\n",
       "      <td>35.73</td>\n",
       "      <td>31.45</td>\n",
       "      <td>42.97</td>\n",
       "      <td>64.94</td>\n",
       "      <td>47.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000391616</td>\n",
       "      <td>ENSG00000240356</td>\n",
       "      <td>RPL23AP7</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>8.72</td>\n",
       "      <td>4.10</td>\n",
       "      <td>6.94</td>\n",
       "      <td>7.45</td>\n",
       "      <td>6.05</td>\n",
       "      <td>1.51</td>\n",
       "      <td>7.12</td>\n",
       "      <td>9.72</td>\n",
       "      <td>2.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000416673</td>\n",
       "      <td>ENSG00000240356</td>\n",
       "      <td>RPL23AP7</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.86</td>\n",
       "      <td>1.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.88</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000334816</td>\n",
       "      <td>ENSG00000186132</td>\n",
       "      <td>C2orf76</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>15.74</td>\n",
       "      <td>12.20</td>\n",
       "      <td>15.45</td>\n",
       "      <td>15.40</td>\n",
       "      <td>7.26</td>\n",
       "      <td>5.28</td>\n",
       "      <td>12.36</td>\n",
       "      <td>18.28</td>\n",
       "      <td>4.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000309451</td>\n",
       "      <td>ENSG00000173272</td>\n",
       "      <td>MZT2A</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>164.77</td>\n",
       "      <td>105.71</td>\n",
       "      <td>107.13</td>\n",
       "      <td>66.56</td>\n",
       "      <td>221.92</td>\n",
       "      <td>37.27</td>\n",
       "      <td>82.36</td>\n",
       "      <td>67.00</td>\n",
       "      <td>50.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id          gene_id gene_name  \\\n",
       "0  ENST00000324913  ENSG00000125611    CHCHD5   \n",
       "1  ENST00000391616  ENSG00000240356  RPL23AP7   \n",
       "2  ENST00000416673  ENSG00000240356  RPL23AP7   \n",
       "3  ENST00000334816  ENSG00000186132   C2orf76   \n",
       "4  ENST00000309451  ENSG00000173272     MZT2A   \n",
       "\n",
       "                       gene_biotype  Brain - Frontal Cortex (BA9)  \\\n",
       "0                    protein_coding                         43.81   \n",
       "1  transcribed_processed_pseudogene                          8.72   \n",
       "2  transcribed_processed_pseudogene                          1.42   \n",
       "3                    protein_coding                         15.74   \n",
       "4                    protein_coding                        164.77   \n",
       "\n",
       "   Brain - Cerebellar Hemisphere  Brain - Putamen (basal ganglia)   Lung  \\\n",
       "0                          19.96                            46.36  83.30   \n",
       "1                           4.10                             6.94   7.45   \n",
       "2                            NaN                              NaN   1.86   \n",
       "3                          12.20                            15.45  15.40   \n",
       "4                         105.71                           107.13  66.56   \n",
       "\n",
       "   Heart - Left Ventricle  Muscle - Skeletal  Heart - Atrial Appendage  \\\n",
       "0                   35.73              31.45                     42.97   \n",
       "1                    6.05               1.51                      7.12   \n",
       "2                    1.03                NaN                       NaN   \n",
       "3                    7.26               5.28                     12.36   \n",
       "4                  221.92              37.27                     82.36   \n",
       "\n",
       "   Cells - Cultured fibroblasts  Liver  \n",
       "0                         64.94  47.60  \n",
       "1                          9.72   2.66  \n",
       "2                          1.88    NaN  \n",
       "3                         18.28   4.89  \n",
       "4                         67.00  50.50  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# table of the median CPM of an isoform in each tissue\n",
    "isoforms_in_tissue.to_csv('../../tables/GTEx_expression/GTEx_isoforms_in_tissues_passing_med_CPM_gt_1.tsv', sep='\\t', index=False)\n",
    "isoforms_in_tissue.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b3f834d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Brain - Frontal Cortex (BA9)</th>\n",
       "      <th>Brain - Cerebellar Hemisphere</th>\n",
       "      <th>Brain - Putamen (basal ganglia)</th>\n",
       "      <th>Lung</th>\n",
       "      <th>Heart - Left Ventricle</th>\n",
       "      <th>Muscle - Skeletal</th>\n",
       "      <th>Heart - Atrial Appendage</th>\n",
       "      <th>Cells - Cultured fibroblasts</th>\n",
       "      <th>Liver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SCYL3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000001036</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>FUCA2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id    gene_biotype gene_name  Brain - Frontal Cortex (BA9)  \\\n",
       "0  ENSG00000000003  protein_coding    TSPAN6                           1.0   \n",
       "1  ENSG00000000005  protein_coding      TNMD                           1.0   \n",
       "2  ENSG00000000419  protein_coding      DPM1                           5.0   \n",
       "3  ENSG00000000457  protein_coding     SCYL3                           1.0   \n",
       "4  ENSG00000001036  protein_coding     FUCA2                           1.0   \n",
       "\n",
       "   Brain - Cerebellar Hemisphere  Brain - Putamen (basal ganglia)  Lung  \\\n",
       "0                            1.0                              1.0   1.0   \n",
       "1                            NaN                              NaN   NaN   \n",
       "2                            4.0                              4.0   5.0   \n",
       "3                            1.0                              1.0   1.0   \n",
       "4                            1.0                              1.0   1.0   \n",
       "\n",
       "   Heart - Left Ventricle  Muscle - Skeletal  Heart - Atrial Appendage  \\\n",
       "0                     1.0                1.0                       1.0   \n",
       "1                     NaN                1.0                       NaN   \n",
       "2                     4.0                5.0                       4.0   \n",
       "3                     1.0                1.0                       1.0   \n",
       "4                     1.0                1.0                       1.0   \n",
       "\n",
       "   Cells - Cultured fibroblasts  Liver  \n",
       "0                           1.0    1.0  \n",
       "1                           NaN    NaN  \n",
       "2                           5.0    3.0  \n",
       "3                           1.0    1.0  \n",
       "4                           1.0    1.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# table of the number of isoforms passing our thresholds (median-unique-counts >= 1 and CPM > 1) per gene\n",
    "n_tx_per_gene_by_tissue.to_csv(\"../../tables/GTEx_expression/GTEx_number_of_tx_per_gene_passing_thresholds_2023.tsv\", sep='\\t', index=False)\n",
    "# tables of the number of protein-coding isoforms passing our thresholds (median-unique-counts >= 1 and CPM > 1) per gene\n",
    "n_tx_per_pc_gene_by_tissue.to_csv(\"../../tables/GTEx_expression/GTEx_number_of_protein_coding_tx_per_gene_passing_thresholds_2023.tsv\", sep='\\t', index=False)\n",
    "n_tx_per_gene_by_tissue.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9a18a6",
   "metadata": {},
   "source": [
    "## Identify the isoforms unique to each tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "56f10aea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_tx</th>\n",
       "      <th>tissue</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3277</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1182</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "      <td>pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>928</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>546</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "      <td>pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>497</td>\n",
       "      <td>Brain - Putamen (basal ganglia)</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   n_tx                           tissue type\n",
       "0  3277    Brain - Cerebellar Hemisphere  all\n",
       "0  1182    Brain - Cerebellar Hemisphere   pc\n",
       "0   928     Brain - Frontal Cortex (BA9)  all\n",
       "0   546     Brain - Frontal Cortex (BA9)   pc\n",
       "0   497  Brain - Putamen (basal ganglia)  all"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'tissue specific' isoforms\n",
    "unique_tiss_isoforms = pd.DataFrame(columns = ['n_tx', 'tissue', 'type'])\n",
    "\n",
    "tiss_spec_iso = {}\n",
    "for tiss in tissues_to_use:\n",
    "    tiss_iso_df = pd.DataFrame(list(sets_of_isoforms_at_cpm_gt_1[tiss]), columns=['isoform'])\n",
    "    tiss_iso_df['tissue'] = tiss\n",
    "    tmp_iso_set = set()\n",
    "    for tiss2 in tissues_to_use:\n",
    "        if tiss != tiss2:\n",
    "            tmp_iso_set.update(sets_of_isoforms_at_cpm_gt_1[tiss2])\n",
    "            \n",
    "    tiss_spec_iso[tiss] = sets_of_isoforms_at_cpm_gt_1[tiss] - tmp_iso_set\n",
    "    unique_tiss_isoforms = pd.concat([unique_tiss_isoforms, pd.DataFrame([{'n_tx': len(tiss_spec_iso[tiss]), 'tissue': tiss, 'type': 'all'}])])\n",
    "    unique_tiss_isoforms = pd.concat([unique_tiss_isoforms, pd.DataFrame([{'n_tx': len(tiss_spec_iso[tiss] & sets_of_pc_isoforms_at_cpm_gt_1[tiss]), 'tissue': tiss, 'type': 'pc'}])])\n",
    "\n",
    "unique_tiss_isoforms.to_csv('../../tables/GTEx_expression/GTEx_tissue_unique_isoforms_by_type_CPM_gt_1.tsv', sep='\\t', index=False)\n",
    "unique_tiss_isoforms.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50293648",
   "metadata": {},
   "source": [
    "## Transcript biotype by tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a4dbc0d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Brain - Frontal Cortex (BA9)\n",
      "Brain - Cerebellar Hemisphere\n",
      "Brain - Putamen (basal ganglia)\n",
      "Lung\n",
      "Heart - Left Ventricle\n",
      "Muscle - Skeletal\n",
      "Heart - Atrial Appendage\n",
      "Cells - Cultured fibroblasts\n",
      "Liver\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>protein_coding</th>\n",
       "      <td>14028</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lncRNA</th>\n",
       "      <td>1257</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>retained_intron</th>\n",
       "      <td>1138</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonsense_mediated_decay</th>\n",
       "      <td>771</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_coding_CDS_not_defined</th>\n",
       "      <td>646</td>\n",
       "      <td>Brain - Frontal Cortex (BA9)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                count                        tissue\n",
       "transcript_biotype                                                 \n",
       "protein_coding                  14028  Brain - Frontal Cortex (BA9)\n",
       "lncRNA                           1257  Brain - Frontal Cortex (BA9)\n",
       "retained_intron                  1138  Brain - Frontal Cortex (BA9)\n",
       "nonsense_mediated_decay           771  Brain - Frontal Cortex (BA9)\n",
       "protein_coding_CDS_not_defined    646  Brain - Frontal Cortex (BA9)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biotype_by_tissue = pd.DataFrame()\n",
    "# showing the top 5 isoform biotypes by tissue\n",
    "for tissue in df_by_tissue:\n",
    "    if tissue in tissues_to_use:\n",
    "        print(tissue)\n",
    "        tmp = df_by_tissue[tissue].copy()\n",
    "        tmp = tmp.loc[tmp[\"median_CPM\"] > 1]\n",
    "        o_t = tmp['transcript_biotype'].value_counts().to_frame().head()\n",
    "        o_t['tissue'] = tissue\n",
    "        biotype_by_tissue = pd.concat([biotype_by_tissue, o_t])\n",
    "\n",
    "biotype_by_tissue.to_csv('../../tables/GTEx_expression/GTEx_transcript_biotype_expr_by_tissue_2023.tsv', sep='\\t')\n",
    "biotype_by_tissue.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da7f3c52",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e68213f3",
   "metadata": {},
   "source": [
    "## Look at the number of isoforms expressed in tissue according to relative abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a82c6efa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Brain - Frontal Cortex (BA9)\n",
      "Brain - Cerebellar Hemisphere\n",
      "Brain - Putamen (basal ganglia)\n",
      "Lung\n",
      "Heart - Left Ventricle\n",
      "Muscle - Skeletal\n",
      "Heart - Atrial Appendage\n",
      "Cells - Cultured fibroblasts\n",
      "Liver\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Brain - Frontal Cortex (BA9)</th>\n",
       "      <th>Brain - Cerebellar Hemisphere</th>\n",
       "      <th>Brain - Putamen (basal ganglia)</th>\n",
       "      <th>Lung</th>\n",
       "      <th>Heart - Left Ventricle</th>\n",
       "      <th>Muscle - Skeletal</th>\n",
       "      <th>Heart - Atrial Appendage</th>\n",
       "      <th>Cells - Cultured fibroblasts</th>\n",
       "      <th>Liver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>22.099448</td>\n",
       "      <td>13.800503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.768104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>33.072917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000324913</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>77.539698</td>\n",
       "      <td>74.569664</td>\n",
       "      <td>81.238979</td>\n",
       "      <td>96.518609</td>\n",
       "      <td>84.485838</td>\n",
       "      <td>74.173987</td>\n",
       "      <td>85.993486</td>\n",
       "      <td>97.209181</td>\n",
       "      <td>98.010974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000391616</td>\n",
       "      <td>ENSG00000240356</td>\n",
       "      <td>59.026128</td>\n",
       "      <td>60.261004</td>\n",
       "      <td>70.042754</td>\n",
       "      <td>47.743446</td>\n",
       "      <td>48.905109</td>\n",
       "      <td>56.273183</td>\n",
       "      <td>64.265537</td>\n",
       "      <td>39.353814</td>\n",
       "      <td>56.624319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000334816</td>\n",
       "      <td>ENSG00000186132</td>\n",
       "      <td>87.331839</td>\n",
       "      <td>79.566825</td>\n",
       "      <td>83.915682</td>\n",
       "      <td>83.648110</td>\n",
       "      <td>80.056180</td>\n",
       "      <td>71.602218</td>\n",
       "      <td>81.610463</td>\n",
       "      <td>80.871886</td>\n",
       "      <td>88.348271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000309451</td>\n",
       "      <td>ENSG00000173272</td>\n",
       "      <td>96.299004</td>\n",
       "      <td>90.897860</td>\n",
       "      <td>95.019477</td>\n",
       "      <td>96.259549</td>\n",
       "      <td>98.601543</td>\n",
       "      <td>94.605768</td>\n",
       "      <td>98.019183</td>\n",
       "      <td>94.972749</td>\n",
       "      <td>96.595499</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id          gene_id  Brain - Frontal Cortex (BA9)  \\\n",
       "0  ENST00000488147  ENSG00000227232                     22.099448   \n",
       "1  ENST00000324913  ENSG00000125611                     77.539698   \n",
       "2  ENST00000391616  ENSG00000240356                     59.026128   \n",
       "3  ENST00000334816  ENSG00000186132                     87.331839   \n",
       "4  ENST00000309451  ENSG00000173272                     96.299004   \n",
       "\n",
       "   Brain - Cerebellar Hemisphere  Brain - Putamen (basal ganglia)       Lung  \\\n",
       "0                      13.800503                              NaN  26.768104   \n",
       "1                      74.569664                        81.238979  96.518609   \n",
       "2                      60.261004                        70.042754  47.743446   \n",
       "3                      79.566825                        83.915682  83.648110   \n",
       "4                      90.897860                        95.019477  96.259549   \n",
       "\n",
       "   Heart - Left Ventricle  Muscle - Skeletal  Heart - Atrial Appendage  \\\n",
       "0                     NaN          50.000000                 33.072917   \n",
       "1               84.485838          74.173987                 85.993486   \n",
       "2               48.905109          56.273183                 64.265537   \n",
       "3               80.056180          71.602218                 81.610463   \n",
       "4               98.601543          94.605768                 98.019183   \n",
       "\n",
       "   Cells - Cultured fibroblasts      Liver  \n",
       "0                           NaN  33.333333  \n",
       "1                     97.209181  98.010974  \n",
       "2                     39.353814  56.624319  \n",
       "3                     80.871886  88.348271  \n",
       "4                     94.972749  96.595499  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter isoforms by median relative abundance instead of CPM and determine the number of isoforms expressed per gene\n",
    "median_rel_ab_iso_in_tis = pd.DataFrame(columns=['gene_id', 'transcript_id'])\n",
    "median_rel_ab_iso_in_tis_n_iso = pd.DataFrame(columns=['gene_id'])\n",
    "\n",
    "# for each tissue\n",
    "for tissue in df_by_tissue:\n",
    "    print(tissue)\n",
    "    tmp = df_by_tissue[tissue].copy()\n",
    "    # isoforms need to account for at least 10 percent of the relative abundance of a gene\n",
    "    tmp = tmp.loc[tmp['median_relative_abundance'] > 10]\n",
    "    tmp_formatted = tmp.copy()[['transcript_id', 'gene_id', 'gene_name', 'median_relative_abundance']].rename(columns={\"median_relative_abundance\": tissue})\n",
    "    median_rel_ab_iso_in_tis = median_rel_ab_iso_in_tis.merge(tmp_formatted, how='outer')\n",
    "    n_iso_tmp = tmp.copy()[[\"gene_id\", \"transcript_id\"]].groupby(\"gene_id\").size().reset_index(name=tissue)\n",
    "    median_rel_ab_iso_in_tis_n_iso = median_rel_ab_iso_in_tis_n_iso.merge(n_iso_tmp, how='outer')\n",
    "\n",
    "# write the isoforms and their median relative abundance to file\n",
    "# write the number of isoforms per gene to file\n",
    "median_rel_ab_iso_in_tis.to_csv('../../tables/GTEx_expression/GTEx_isoforms_in_tissues_passing_median_rel_abund_gt_10.tsv', sep='\\t', index=False)\n",
    "median_rel_ab_iso_in_tis_n_iso.to_csv('../../tables/GTEx_expression/GTEx_number_of_tx_per_gene_passing_median_rel_abund_gt_10_threshold_2023.tsv', sep='\\t', index=False)\n",
    "median_rel_ab_iso_in_tis.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
