{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b40cead4",
   "metadata": {},
   "source": [
    "# Library Import and Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8ae625f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "## Display all rows of pandas dataframes\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f650c5dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "name: relative_transcript_abundance\n",
    "\n",
    "purpose: calculate relative transcript abundance\n",
    "\n",
    "input: a dataframe with a ref_gene_id column identifying the transcript gene of origin and a cov columns with \n",
    "the coverage for the transcripts.\n",
    "\n",
    "output: the same dataframe with median relative abundance and mean gene CPM columns added\n",
    "'''\n",
    "\n",
    "def relative_transcript_abundance(df):\n",
    "    \n",
    "    first=True\n",
    "    \n",
    "    dff = df.copy()\n",
    "\n",
    "    # for each sample's CPM column, calculate the relative abundance of the transcript in the gene and create a new column with that value\n",
    "    for col in dff.filter(regex='bam_CPM').columns:\n",
    "        \n",
    "        col_gene_name = col.split(\"_CP\")[0] + \"_total_gene_CPM\"\n",
    "        col_relative_abundance = col.split(\"_CP\")[0] + \"_relative_abundance\"\n",
    "    \n",
    "        dff_sums = dff[[\"gene_id\", col]].groupby(\"gene_id\").sum()\n",
    "\n",
    "        dff_sums[col_gene_name] = dff_sums[col].copy()\n",
    "\n",
    "        dff_sums.drop(columns=col, inplace=True)\n",
    "\n",
    "        if first:\n",
    "            merged_dff = pd.merge(dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "            merged_dff[col_relative_abundance] = ((merged_dff[col]/merged_dff[col_gene_name]) * 100)\n",
    "            \n",
    "        else:\n",
    "            merged_dff = pd.merge(merged_dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "            merged_dff[col_relative_abundance] = ((merged_dff[col]/merged_dff[col_gene_name]) * 100)\n",
    "        \n",
    "        first=False\n",
    "        \n",
    "    for col_count in merged_dff.filter(regex='bam_count').columns:\n",
    "        \n",
    "        col_gene_name = col_count.split(\"_count\")[0] + \"_total_gene_counts\"\n",
    "        \n",
    "        dff_sums = merged_dff[[\"gene_id\", col_count]].groupby(\"gene_id\").sum()\n",
    "        dff_sums[col_gene_name] = dff_sums[col_count].copy()\n",
    "    \n",
    "        \n",
    "        dff_sums.drop(columns=col_count, inplace=True)\n",
    "        \n",
    "        merged_dff = pd.merge(merged_dff, dff_sums, how='inner', on=\"gene_id\")\n",
    "        \n",
    "    merged_dff.fillna(value=0, inplace=True)\n",
    "    \n",
    "    # calculate the median relative abundance for all the samples and drop the sample specific relative abundance columns\n",
    "    rel_ab_col = merged_dff.filter(regex='bam_relative_abundance').columns\n",
    "    merged_dff['median_relative_abundance'] = merged_dff[rel_ab_col].copy().median(axis=1)\n",
    "    merged_dff.drop(columns = rel_ab_col, inplace = True)\n",
    "    \n",
    "    # calculate the median gene CPM and drop sample specific gene CPM\n",
    "    gene_cpm_col = merged_dff.filter(regex='bam_total_gene_CPM').columns\n",
    "    merged_dff['median_gene_cpm'] = merged_dff[gene_cpm_col].copy().median(axis=1)\n",
    "    merged_dff.drop(columns = gene_cpm_col, inplace = True)\n",
    "        \n",
    "    return merged_dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a6446e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: fix_column_names\n",
    "\n",
    "purpose: Fixing the column names, making them smaller, informative, and consistent\n",
    "\n",
    "input: The raw counts dataframe for either genes or transcripts \n",
    "\n",
    "output: Same dataframe with improved column names\n",
    "'''\n",
    "\n",
    "def fix_column_names(df, is_gene=False):\n",
    "    \n",
    "    ## Check if this is a gene counts object\n",
    "    if is_gene:\n",
    "        \n",
    "        ## Get count column names and create list of new column names\n",
    "        count_columns = df.columns.tolist()\n",
    "        list_new_names = [\"gene_id\"]\n",
    "        \n",
    "        ## gene_id comes in as index for gene counts data, make it into the first column instead\n",
    "        df[\"gene_id\"] = df.index\n",
    "        cols = list(df.columns)\n",
    "        cols = [cols[-1]] + cols[:-1]\n",
    "        df = df[cols]\n",
    "        df.reset_index(inplace=True, drop=True)\n",
    "    \n",
    "    ## If it is a transcript dataset\n",
    "    else:\n",
    "        ## Set count columns and create list of new names\n",
    "        count_columns = df.columns[2:].tolist()\n",
    "        list_new_names = [ \"transcript_id\", \"gene_id\"]\n",
    "    \n",
    "    ## Fix names one by one and add to list of new names\n",
    "    for col in count_columns:\n",
    "        col = col.split(\"_mapped\")[0] + \"_counts\"\n",
    "        list_new_names.append(col)\n",
    "    \n",
    "    ## Rename columns\n",
    "    df.columns = list_new_names\n",
    "    \n",
    "    return df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "143176a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: parse_df_columns\n",
    "\n",
    "purpose: parsing the last aggregate column of the gtf/gff3 into useful columns and cleaning non-relevant columns\n",
    "\n",
    "input: dataframe containining \"raw\" gtf/gff\n",
    "\n",
    "output: dataframe containing gtf with useful columns [\"gene_id\", \"transcript_id\", etc...]\n",
    "'''\n",
    "\n",
    "def parse_df_columns(df, is_ref=True, is_transcript=False, is_prot=False):\n",
    "\n",
    "    if is_ref:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split(\"ne_id \\\"\", expand=True)[1].str.split('\\\";', expand=True)[0]\n",
    "        \n",
    "        ## Get gene names\n",
    "        df[\"gene_name\"] = df[\"other\"].str.split(\"gene_name \\\"\", expand=True)[1].str.split('\\\";', expand=True)[0]\n",
    "        \n",
    "        ## Get get transcript biotype\n",
    "        df[\"gene_biotype\"] = df[\"other\"].str.split('gene_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "        \n",
    "        ## If is transcript get transcript id and transcript biotype\n",
    "        if is_transcript:\n",
    "            df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            df[\"transcript_biotype\"] = df[\"other\"].str.split('transcript_biotype \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "            \n",
    "            ## If is prot get protein_id\n",
    "            if is_prot:\n",
    "                df[\"protein_id\"] = df[\"other\"].str.split('protein_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "                df[\"ccds_id\"] = df[\"other\"].str.split('ccds_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "                df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        ## Get gene ids\n",
    "        df[\"gene_id\"] = df[\"other\"].str.split('\";', expand=True)[0].str.extract(\"([^ \\\"]*$)\", expand=True)\n",
    "\n",
    "        ## Get transcript ids\n",
    "        df[\"transcript_id\"] = df[\"other\"].str.split('transcript_id \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Get exon number\n",
    "        df[\"exon_number\"] = df[\"other\"].str.split('exon_number \"', expand=True)[1].str.split('\"', expand=True)[0]\n",
    "\n",
    "        ## Drop \"other\" column\n",
    "        df.drop(columns=[\"other\", \"dot_1\", \"dot_2\"], inplace=True)\n",
    "\n",
    "    for col in df.columns:\n",
    "        df.loc[df[col].isnull(), col] = np.NaN\n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1117a08",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: calculate_cpm\n",
    "\n",
    "purpose: Calculate CPM for the each sample given\n",
    "\n",
    "input: Counts dataset\n",
    "\n",
    "output: Counts dataset with CPM columns as well\n",
    "'''\n",
    "\n",
    "def calculate_cpm(df, is_gene=False):\n",
    "\n",
    "    ## Set count columns if dataframe is gene counts\n",
    "    if is_gene:\n",
    "        count_columns = df.columns[1:].tolist()\n",
    "    \n",
    "    ## Set count columns if dataframe is transcript counts\n",
    "    else:\n",
    "        count_columns = df.columns[2:].tolist()\n",
    "\n",
    "    ## Loop through counts columns to calculate CPM and add to the dataframe\n",
    "    for col in count_columns:\n",
    "        \n",
    "        df[col] = round(df[col], 2)\n",
    "        cpm_name = col.replace(\"_counts\", \"_CPM\")\n",
    "        df[cpm_name] = round(((df[col]/(df[col].sum())) * 1000000), 2)\n",
    "    \n",
    "    return df  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "28b9546f",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function name: get_samp_id\n",
    "\n",
    "purpose: \n",
    "\n",
    "input: sample string name\n",
    "\n",
    "output: sample string name with bam appended\n",
    "'''\n",
    "\n",
    "\n",
    "def get_samp_id(s):\n",
    "    \n",
    "    name_array = s.split('.')\n",
    "    last_element = name_array[-1].split('_')[0]\n",
    "    string_1 = '-'.join(name_array[:-1])\n",
    "    string_2 = f\"{last_element}.bam\"\n",
    "    return f\"{string_1}.{string_2}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "21852d06",
   "metadata": {},
   "outputs": [],
   "source": [
    "## define tissues we are using\n",
    "tissues_to_use = [\"Brain - Cerebellar Hemisphere\",\n",
    "                   \"Brain - Frontal Cortex (BA9)\",\n",
    "                   \"Brain - Putamen (basal ganglia)\",\n",
    "                   \"Cells - Cultured fibroblasts\",\n",
    "                   \"Heart - Atrial Appendage\",\n",
    "                   \"Heart - Left Ventricle\",\n",
    "                   \"Liver\",\n",
    "                   \"Lung\",\n",
    "                   \"Muscle - Skeletal\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19eb60d8",
   "metadata": {},
   "source": [
    "# Expression of new transcripts since 2019 in our Long-read dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c4e5a7d",
   "metadata": {},
   "source": [
    "## Import 2023 reference and disease gene lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dfa4bd6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open original reference\n",
    "ref_2023 = pd.read_csv(\"../../references/Homo_sapiens.GRCh38.109.gtf\", header=None, delimiter=\"\\t\", low_memory=False, \n",
    "                       names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"], comment=\"#\")\n",
    "\n",
    "# filter out anything with ERCC (shouldn't be there, but just in case)\n",
    "ref_2023 = ref_2023.loc[~ref_2023[\"chr\"].str.startswith(\"ERCC-\")]\n",
    "\n",
    "## Parse through reference to get gene names and ids\n",
    "gene_ref_2023 = ref_2023.loc[ref_2023[\"type\"]==\"gene\"].copy()\n",
    "gene_ref_2023 = parse_df_columns(gene_ref_2023, is_ref=True)\n",
    "gene_info = gene_ref_2023[['gene_id', 'gene_biotype', 'gene_name']].copy()\n",
    "\n",
    "## Create 2023 transcript ref\n",
    "transcript_ref_2023 = ref_2023.loc[ref_2023[\"type\"]==\"transcript\"].copy()\n",
    "transcript_ref_2023 = parse_df_columns(transcript_ref_2023, is_ref=True, is_transcript=True)\n",
    "# get the transcript biotypes\n",
    "transcript_biotype = transcript_ref_2023[['transcript_id', 'transcript_biotype']].copy()\n",
    "\n",
    "## Import disease relevant genes\n",
    "disease_relevant_genes = pd.read_csv(\"../../references/medically_relevant_genes.tsv\", sep=\"\\t\")\n",
    "\n",
    "## Brain disease genes\n",
    "brain_disease_gene_ids = pd.read_csv(\"../../references/brain_disease_genes_only_IDs.tsv\", sep=\"\\t\")\n",
    "brain_disease_annotations = pd.read_csv(\"../../references/brain_disease_genes_with_disease.tsv\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "## Create disease relevant list including chromosome\n",
    "disease_relevant_genes_annotated = disease_relevant_genes.merge(gene_ref_2023[[\"gene_id\", \"gene_name\", \"chr\"]], \n",
    "                                                               how=\"inner\", on=[\"gene_id\", \"gene_name\"])\n",
    "\n",
    "## Create list of protein coding genes\n",
    "protein_coding_ref_2023 = gene_ref_2023.loc[gene_ref_2023[\"gene_biotype\"] == \"protein_coding\"].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26dbd6b6",
   "metadata": {},
   "source": [
    "## Import 2019 reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d7286770",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Open original reference\n",
    "ref_2019 = pd.read_csv(\"../../references/Homo_sapiens.GRCh38.96.gtf\", header=None, delimiter=\"\\t\", low_memory=False, \n",
    "                       names=[\"chr\", \"source\", \"type\", \"start\", \"end\", \"dot_1\", \"strand\", \"dot_2\", \"other\"], comment=\"#\")\n",
    "\n",
    "# filter out anything with ERCC (shouldn't be there, but just in case)\n",
    "ref_2019 = ref_2019.loc[~ref_2019[\"chr\"].str.startswith(\"ERCC-\")]\n",
    "\n",
    "## Parse through reference to get gene names and ids\n",
    "gene_ref_2019 = ref_2019.loc[ref_2019[\"type\"]==\"gene\"].copy()\n",
    "gene_ref_2019 = parse_df_columns(gene_ref_2019, is_ref=True)\n",
    "gene_info_2019 = gene_ref_2019[['gene_id', 'gene_biotype', 'gene_name']].copy()\n",
    "\n",
    "## Create 2023 transcript ref\n",
    "transcript_ref_2019 = ref_2019.loc[ref_2019[\"type\"]==\"transcript\"].copy()\n",
    "transcript_ref_2019 = parse_df_columns(transcript_ref_2019, is_ref=True, is_transcript=True)\n",
    "transcript_biotype_2019 = transcript_ref_2023[['transcript_id', 'transcript_biotype']].copy()\n",
    "transcript_ref_2019_ids = transcript_ref_2019[\"transcript_id\"].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80234ded",
   "metadata": {},
   "source": [
    "## Load reference for protein coding sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "20492add",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_cds = pd.read_csv(\"../../tables/transcript_id_and_unique_protein_id_reference.tsv\", delimiter=\"\\t\", low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e96bcef0",
   "metadata": {},
   "source": [
    "## Load transcript counts matrix and CPM normalize it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a7026ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# samples we are not using due to replicates, low read depth, and poor PCA clustering\n",
    "filter_out_samples = [ \n",
    "    \"GTEX-Q2AG-0011-R11A-SM-2EBL2_rep2.FAK44637.bam\", \n",
    "    \"GTEX-Q2AG-0011-R11A-SM-2EBL2_rep.FAK49243.bam\", \n",
    "    \"GTEX-T5JC-0011-R10A-SM-2TT23.FAK91589.bam\", \n",
    "    \"GTEX-QEG5-0008-SM-3QHW2_exp.FAK30166.bam\", \n",
    "    \"GTEX-QV44-0008-SM-3QNG7_ctrl1.FAK55556.bam\",\n",
    "    \"GTEX-QV44-0008-SM-3QNG7_exp.FAK52124.bam\",\n",
    "    \"GTEX-RWS6-0008-SM-3QHWG_rep.FAK49207.bam\", \n",
    "    \"GTEX-S4Z8-0008-SM-2Y983_exp1.FAK55723.bam\",\n",
    "    \"GTEX-S4Z8-0008-SM-2Y983_exp2.FAK47416.bam\",\n",
    "    \"GTEX-S95S-0008-SM-3RQ8B_exp1.FAK55217.bam\",\n",
    "    \"GTEX-S95S-0008-SM-3RQ8B_exp2.FAK47088.bam\",\n",
    "    \"GTEX-WY7C-0008-SM-3NZB5_ctrl.FAK55679.bam\",\n",
    "    \"GTEX-1GN1W-0226-SM-7AGLJ_rep.FAK91654.bam\",\n",
    "    \"GTEX-WY7C-1126-SM-3GS2X_rep2.FAK49168.bam\",\n",
    "    \"GTEX-WY7C-1126-SM-3GS2X.FAK39149.bam\",\n",
    "    \"GTEX-Y5LM-0426-SM-3YX99.FAK52212.bam\",\n",
    "    \"GTEX-Y5LM-0426-SM-3YX99_rep2.FAK41279.bam\",\n",
    "    \"GTEX-14BMU-0526-SM-5CA2F.FAK44778.bam\",\n",
    "    \"GTEX-14BMU-0526-SM-5CA2F_rep.FAK93376.bam\",\n",
    "    \"GTEX-13QJ3-0726-SM-7LDHS.FAK49189.bam\",\n",
    "    \"GTEX-ZT9X-1826-SM-4V2KV_rep.FAK39773.bam\",\n",
    "    \"GTEX-ZT9X-1826-SM-4V2KV.FAK49260.bam\", \n",
    "    \"GTEX-WY7C-0726-SM-3GLGQ.FAK46872.bam\" ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "83809e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load matrix\n",
    "df = pd.read_csv(\"../../data/raw/GRCh38_quant_mapq10_gtf_109_and_high-confidence_GTEx_DATA/bambu_quant/counts_transcript.txt\", sep=\"\\t\")\n",
    "# load unique counts matrix\n",
    "df_unique = pd.read_csv(\"../../data/raw/GRCh38_quant_mapq10_gtf_109_and_high-confidence_GTEx_DATA/bambu_quant/uniqueCounts_transcript.txt\", sep=\"\\t\")\n",
    "# load gtex metadata\n",
    "gtex_metadata = pd.read_csv(\"/pscratch/mteb223_uksr/new_RNA_isoform_expression_across_tissues/data/GTEx_v9_ONT_metadata.txt\", sep = \"\\t\")\n",
    "\n",
    "## Convert sample names to bam files\n",
    "columns_to_rename = df.columns[2:]\n",
    "new_col_names = {col: get_samp_id(col) for col in columns_to_rename}\n",
    "df.rename(columns=new_col_names, inplace=True)\n",
    "df = df[df.columns.difference(filter_out_samples)]\n",
    "df_unique.rename(columns=new_col_names, inplace=True)\n",
    "df_unique = df_unique[df_unique.columns.difference(filter_out_samples)]\n",
    "\n",
    "# get the list of samples so that we can separate out the columns below by tissue\n",
    "df_cols_dict = {}\n",
    "for tissue in gtex_metadata[\"tissue_site_detail\"].copy().unique():\n",
    "    for tissue in tissues_to_use:\n",
    "        df_cols_dict[tissue] = [el for el in gtex_metadata.loc[gtex_metadata['tissue_site_detail'] == tissue, 'bam_file'].values.tolist() if not 'direct' in el]\n",
    "        df_cols_dict[tissue] = [x for x in df_cols_dict[tissue] if x not in filter_out_samples]\n",
    "\n",
    "df_by_tissue = {}\n",
    "#print(df.head())\n",
    "for key in df_cols_dict:\n",
    "\n",
    "    # grab the columns we will need for this tissue\n",
    "    columns = [\"TXNAME\", \"GENEID\"] + df_cols_dict[key]\n",
    "    tmp = df[columns].copy()\n",
    "    \n",
    "    ## Fix columns names\n",
    "    tmp = fix_column_names(tmp, is_gene=False)\n",
    "    \n",
    "    ## Add total counts column\n",
    "    tmp[\"total_counts\"] = tmp[tmp.filter(regex='count').columns].copy().sum(axis=1)\n",
    "    \n",
    "    ## CPM normalize counts matrix\n",
    "    tmp = calculate_cpm(tmp, is_gene=False)\n",
    "    tmp[\"median_CPM\"] = round(tmp[tmp.filter(regex='CPM').columns].copy().median(axis=1), 2)\n",
    "    tmp.drop(columns=tmp.filter(regex='count').columns, inplace=True)\n",
    "    \n",
    "    tmp_u = df_unique[columns].copy()\n",
    "    \n",
    "    ## Fix columns names\n",
    "    tmp_u = fix_column_names(tmp_u, is_gene=False)\n",
    "    ## Add total counts column\n",
    "    tmp_u[\"median_unique_counts\"] = tmp_u[tmp_u.filter(regex='count').columns].copy().median(axis=1)\n",
    "    # for the unique dataframe, keep only these columns for merging\n",
    "    tmp_u = tmp_u[['transcript_id', 'gene_id', 'median_unique_counts']]\n",
    "\n",
    "    # merge the all counts and unique counts together\n",
    "    tmp = pd.merge(tmp, tmp_u, how='inner')\n",
    "    # calculate the relative transcript abundance\n",
    "    tmp = relative_transcript_abundance(tmp.copy())\n",
    "    # add the tissue name\n",
    "    tmp['tissue_site_detail'] = key\n",
    "    # add on the biotype of the transcript and gene\n",
    "    tmp = tmp.merge(transcript_biotype, how='left')\n",
    "    tmp = tmp.merge(gene_info, how='left')\n",
    "    # write table to file\n",
    "    tmp.to_csv(\"../../data/processed/GTEx/GTEX_since2019_\" + key + \"_cpm_transcript.txt\", sep=\"\\t\", index=False)\n",
    "    n_filtered_out = len(tmp.loc[tmp[\"median_unique_counts\"] < 1])\n",
    "    # filter -> transcripts must be present with at least one unique count in at least half the samples\n",
    "    tmp = tmp.loc[tmp[\"median_unique_counts\"] >= 1].copy()\n",
    "    df_by_tissue[key] = tmp\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "43b7c3ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>GTEX-1211K-0826-SM-7LDFQ.FAK46515.bam_CPM</th>\n",
       "      <th>GTEX-14BMU-0526-SM-5CA2F_rep2.FAK49039.bam_CPM</th>\n",
       "      <th>GTEX-1I6K7-1226-SM-AAEQX.FAK44642.bam_CPM</th>\n",
       "      <th>GTEX-1KXAM-0426-SM-CYKMP.FAK44752.bam_CPM</th>\n",
       "      <th>GTEX-WYVS-0526-SM-3H5V7.FAK54827.bam_CPM</th>\n",
       "      <th>GTEX-ZT9X-0326-SM-4U9QG.FAK44894.bam_CPM</th>\n",
       "      <th>total_CPM</th>\n",
       "      <th>median_CPM</th>\n",
       "      <th>median_unique_counts</th>\n",
       "      <th>median_relative_abundance</th>\n",
       "      <th>median_gene_cpm</th>\n",
       "      <th>tissue_site_detail</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BambuTx1</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>0.45</td>\n",
       "      <td>1.67</td>\n",
       "      <td>3.33</td>\n",
       "      <td>2.07</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.31</td>\n",
       "      <td>1.56</td>\n",
       "      <td>1.56</td>\n",
       "      <td>6.5</td>\n",
       "      <td>73.231896</td>\n",
       "      <td>2.600</td>\n",
       "      <td>Lung</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000488147</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.04</td>\n",
       "      <td>0.37</td>\n",
       "      <td>1.83</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2.0</td>\n",
       "      <td>26.768104</td>\n",
       "      <td>2.600</td>\n",
       "      <td>Lung</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BambuTx100</td>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>3.30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.96</td>\n",
       "      <td>3.19</td>\n",
       "      <td>2.72</td>\n",
       "      <td>3.22</td>\n",
       "      <td>2.27</td>\n",
       "      <td>2.72</td>\n",
       "      <td>9.5</td>\n",
       "      <td>15.137348</td>\n",
       "      <td>19.685</td>\n",
       "      <td>Lung</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BambuTx97</td>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>11.94</td>\n",
       "      <td>21.86</td>\n",
       "      <td>22.99</td>\n",
       "      <td>22.50</td>\n",
       "      <td>9.86</td>\n",
       "      <td>12.88</td>\n",
       "      <td>17.01</td>\n",
       "      <td>17.01</td>\n",
       "      <td>41.5</td>\n",
       "      <td>79.024450</td>\n",
       "      <td>19.685</td>\n",
       "      <td>Lung</td>\n",
       "      <td>NaN</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BambuTx1006</td>\n",
       "      <td>ENSG00000125611</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.26</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.28</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.324856</td>\n",
       "      <td>86.400</td>\n",
       "      <td>Lung</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CHCHD5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     transcript_id          gene_id  \\\n",
       "0         BambuTx1  ENSG00000227232   \n",
       "1  ENST00000488147  ENSG00000227232   \n",
       "2       BambuTx100  ENSG00000215861   \n",
       "4        BambuTx97  ENSG00000215861   \n",
       "8      BambuTx1006  ENSG00000125611   \n",
       "\n",
       "   GTEX-1211K-0826-SM-7LDFQ.FAK46515.bam_CPM  \\\n",
       "0                                       0.45   \n",
       "1                                       0.00   \n",
       "2                                       3.30   \n",
       "4                                      11.94   \n",
       "8                                       0.00   \n",
       "\n",
       "   GTEX-14BMU-0526-SM-5CA2F_rep2.FAK49039.bam_CPM  \\\n",
       "0                                            1.67   \n",
       "1                                            0.42   \n",
       "2                                            0.00   \n",
       "4                                           21.86   \n",
       "8                                            0.84   \n",
       "\n",
       "   GTEX-1I6K7-1226-SM-AAEQX.FAK44642.bam_CPM  \\\n",
       "0                                       3.33   \n",
       "1                                       0.56   \n",
       "2                                       1.96   \n",
       "4                                      22.99   \n",
       "8                                       0.28   \n",
       "\n",
       "   GTEX-1KXAM-0426-SM-CYKMP.FAK44752.bam_CPM  \\\n",
       "0                                       2.07   \n",
       "1                                       1.04   \n",
       "2                                       3.19   \n",
       "4                                      22.50   \n",
       "8                                       0.26   \n",
       "\n",
       "   GTEX-WYVS-0526-SM-3H5V7.FAK54827.bam_CPM  \\\n",
       "0                                      0.56   \n",
       "1                                      0.37   \n",
       "2                                      2.72   \n",
       "4                                      9.86   \n",
       "8                                      0.19   \n",
       "\n",
       "   GTEX-ZT9X-0326-SM-4U9QG.FAK44894.bam_CPM  total_CPM  median_CPM  \\\n",
       "0                                      1.31       1.56        1.56   \n",
       "1                                      1.83       0.72        0.56   \n",
       "2                                      3.22       2.27        2.72   \n",
       "4                                     12.88      17.01       17.01   \n",
       "8                                      0.78       0.42        0.28   \n",
       "\n",
       "   median_unique_counts  median_relative_abundance  median_gene_cpm  \\\n",
       "0                   6.5                  73.231896            2.600   \n",
       "1                   2.0                  26.768104            2.600   \n",
       "2                   9.5                  15.137348           19.685   \n",
       "4                  41.5                  79.024450           19.685   \n",
       "8                   1.0                   0.324856           86.400   \n",
       "\n",
       "  tissue_site_detail      transcript_biotype  \\\n",
       "0               Lung                     NaN   \n",
       "1               Lung  unprocessed_pseudogene   \n",
       "2               Lung                     NaN   \n",
       "4               Lung                     NaN   \n",
       "8               Lung                     NaN   \n",
       "\n",
       "                         gene_biotype gene_name  \n",
       "0              unprocessed_pseudogene    WASH7P  \n",
       "1              unprocessed_pseudogene    WASH7P  \n",
       "2  transcribed_unprocessed_pseudogene       NaN  \n",
       "4  transcribed_unprocessed_pseudogene       NaN  \n",
       "8                      protein_coding    CHCHD5  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_by_tissue['Lung'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "99643dd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Brain - Cerebellar Hemisphere\n",
      "['GTEX-11H98-0011-R11b-SM-4SFLZ.FAK46829.bam', 'GTEX-13VXU-0011-R11b-SM-5BFQZ.FAK44611.bam', 'GTEX-17F97-0011-R11b-SM-63KY2.FAK41775.bam', 'GTEX-1H3NZ-0011-R11b-SM-AUNOV.FAK49024.bam', 'GTEX-Q2AG-0011-R11A-SM-2EBL2.FAK42265.bam', 'GTEX-T5JC-0011-R11A-SM-2TT24.FAK54838.bam']\n",
      "6\n",
      "Brain - Frontal Cortex (BA9)\n",
      "['GTEX-1192X-0011-R10a-SM-4RXXZ.FAK49046.bam', 'GTEX-13X6J-0011-R10b-SM-5CEKT.FAK44896.bam', 'GTEX-14BIL-0011-R10a-SM-5EQV4.FAK49209.bam', 'GTEX-15DCD-0011-R10b-SM-5S51M.FAK42101.bam', 'GTEX-QDT8-0011-R10A-SM-2FKJB.FAK49182.bam']\n",
      "5\n",
      "Brain - Putamen (basal ganglia)\n",
      "['GTEX-11TTK-0011-R7b-SM-4TVFS.FAK39197.bam', 'GTEX-1313W-0011-R7b-SM-4ZL3U.FAK44754.bam', 'GTEX-13RTJ-0011-R7b-SM-5CTCB.FAK49257.bam', 'GTEX-14C5O-0011-R7b-SM-5GUPO.FAK54887.bam', 'GTEX-15ER7-0011-R7a-SM-5QYP2.FAK44704.bam', 'GTEX-T5JC-0011-R7A-SM-2TT1Z.FAK42170.bam']\n",
      "6\n",
      "Cells - Cultured fibroblasts\n",
      "['GTEX-OIZI-0008-SM-2FR3P.FAK41829.bam', 'GTEX-OXRL-0008-SM-2FR3T.FAK39703.bam', 'GTEX-PSDG-0008-SM-3NZBI.FAK41823.bam', 'GTEX-QEG5-0008-SM-3QHW2_ctrl.FAK55175.bam', 'GTEX-QV44-0008-SM-3QNG7_ctrl2.FAK49435.bam', 'GTEX-R55E-0008-SM-3QNG3.FAK46687.bam', 'GTEX-RWS6-0008-SM-3QHWG.FAK46867.bam', 'GTEX-S4Z8-0008-SM-2Y983_ctrl.FAK54991.bam', 'GTEX-S95S-0008-SM-3RQ8B_ctrl.FAK55194.bam', 'GTEX-WFG7-0008-SM-3ML7A.FAK41398.bam', 'GTEX-WY7C-0008-SM-3NZB5_ctrl.FAK55628.bam']\n",
      "11\n",
      "Heart - Atrial Appendage\n",
      "['GTEX-14XAO-0926-SM-5YQWC.FAK46869.bam', 'GTEX-1GN1W-0226-SM-7AGLJ.FAK44655.bam', 'GTEX-1HBPH-0226-SM-7LLUW.FAK41473.bam', 'GTEX-1IDJD-0226-SM-AML89.FAK49187.bam', 'GTEX-1IDJF-0826-SM-ABADH.FAK39145.bam', 'GTEX-WY7C-1126-SM-3GS2X_rep.FAK49218.bam']\n",
      "6\n",
      "Heart - Left Ventricle\n",
      "['GTEX-13QBU-0426-SM-5A4VT.FAK49161.bam', 'GTEX-15RIE-1726-SM-5V7WB.FAK41916.bam', 'GTEX-1H11D-1226-SM-9P6XJ.FAK44759.bam', 'GTEX-1I6K7-0626-SM-AAEQW.FAK49296.bam', 'GTEX-1IDJU-0726-SM-BXZK2.FAK49131.bam', 'GTEX-OHPL-0326-SM-DPSB8.FAK54495.bam', 'GTEX-ZVZP-0226-SM-4VEIO.FAK41904.bam']\n",
      "7\n",
      "Liver\n",
      "['GTEX-R53T-0326-SM-2K8S4.FAK44579.bam', 'GTEX-UTHO-2426-SM-38ZXF.FAK46748.bam', 'GTEX-Y5LM-0426-SM-3YX99_rep.FAK49269.bam', 'GTEX-ZF29-2026-SM-4TVXH.FAK44621.bam', 'GTEX-ZPU1-0826-SM-4UJSC.FAK41797.bam']\n",
      "5\n",
      "Lung\n",
      "['GTEX-1211K-0826-SM-7LDFQ.FAK46515.bam', 'GTEX-14BMU-0526-SM-5CA2F_rep2.FAK49039.bam', 'GTEX-1I6K7-1226-SM-AAEQX.FAK44642.bam', 'GTEX-1KXAM-0426-SM-CYKMP.FAK44752.bam', 'GTEX-WYVS-0526-SM-3H5V7.FAK54827.bam', 'GTEX-ZT9X-0326-SM-4U9QG.FAK44894.bam']\n",
      "6\n",
      "Muscle - Skeletal\n",
      "['GTEX-13QJ3-0726-SM-7LDHS_rep.FAK93371.bam', 'GTEX-17MFQ-1926-SM-6659L.FAK44757.bam', 'GTEX-1C64N-0326-SM-6ZHT1.FAK53752.bam', 'GTEX-1KXAM-2426-SM-CYKMT.FAK39676.bam', 'GTEX-1LVA9-0326-SM-CYRXA.FAK39191.bam', 'GTEX-ZT9X-1826-SM-4V2KV_rep2.FAK44613.bam']\n",
      "6\n",
      "58\n"
     ]
    }
   ],
   "source": [
    "# verify the number of samples being used from each tissue\n",
    "samps_we_use = []\n",
    "total_samples = 0\n",
    "for tiss in df_cols_dict:\n",
    "    # print the tissue, the sample ids, and the number of samples\n",
    "    print(tiss)\n",
    "    print(df_cols_dict[tiss])\n",
    "    samps_we_use = samps_we_use + df_cols_dict[tiss]\n",
    "    print(len(df_cols_dict[tiss]))\n",
    "    total_samples = total_samples + len(df_cols_dict[tiss])\n",
    "print(total_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e817c6b",
   "metadata": {},
   "source": [
    "## Make figures looking at brain level expression of new isoforms since 2019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8e70812f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Only keep transcripts that are exclusive to the 2023 annotation when compared to 2019\n",
    "exclusive_2023_transcripts = transcript_ref_2023.loc[~transcript_ref_2023[\"transcript_id\"].isin(transcript_ref_2019_ids)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e23a0d8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create list of new CCDS since 2019\n",
    "\n",
    "## Get all transcript IDs from ENSEMBL 96 GTF (2019) and their protein_id that we generated\n",
    "cds_2019_ref = transcript_ref_2019.merge(ref_cds, how=\"inner\", on=\"transcript_id\").dropna(subset=\"protein_id\")\n",
    "cds_2019_ref_protein_ids = cds_2019_ref[\"protein_id\"].copy()\n",
    "\n",
    "## Get transcript names from ENSEMBL 109 GTF (2023)\n",
    "cds_2023_ref = transcript_ref_2023.merge(ref_cds, how=\"inner\", on=\"transcript_id\").dropna(subset=\"protein_id\")\n",
    "\n",
    "## Get ensmbl 109 (2023) exclusive protein coding sequences\n",
    "exclusive_2023_cds = cds_2023_ref.loc[~cds_2023_ref[\"transcript_id\"].isin(transcript_ref_2019_ids)].copy()\n",
    "exclusive_2023_cds = exclusive_2023_cds.loc[~exclusive_2023_cds[\"protein_id\"].isin(cds_2019_ref_protein_ids)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c7c6156e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize variables\n",
    "df_exclusive_2023_transcripts = {}\n",
    "df_exclusive_2023_ccds = {}\n",
    "\n",
    "list_cpm_thresh = [x/100 for x in range(0,1002)]\n",
    "\n",
    "## Create lists with numbers of transcripts expressed across CPM thresholds for different categories\n",
    "list_2023_all_transcript_median = {}\n",
    "list_2023_all_gene_median = {}\n",
    "list_2023_new_cds_transcript_median = {}\n",
    "list_2023_med_relevant_transcript_median = {}\n",
    "list_2023_med_relevant_new_cds_transcript_median = {}\n",
    "list_2023_brain_relevant_transcript_median = {}\n",
    "list_2023_brain_relevant_new_cds_transcript_median = {}\n",
    "\n",
    "exclusive_2023_tx_df = pd.DataFrame()\n",
    "isoforms_in_categories = {}\n",
    "\n",
    "values_at_cpm_threshold = pd.DataFrame(columns=['cpm_threshold', 'tissue', \n",
    "                                                'all_transcript_median', \n",
    "                                                'all_gene_median',\n",
    "                                                'new_cds_transcript_median', \n",
    "                                                'med_relevant_transcript_median', \n",
    "                                               'med_relevant_new_cds_transcript_median',\n",
    "                                               'brain_relevant_transcript_median',\n",
    "                                               'brain_relevant_new_cds_transcript_median'])\n",
    "\n",
    "## Create lists with numbers of 2023 exclusive transcripts expressed across CPM thresholds\n",
    "for tissue in df_by_tissue:\n",
    "    \n",
    "    # grab the tissue we are working with\n",
    "    tmp = df_by_tissue[tissue].copy()\n",
    "    # filter to only new since 2019 TX\n",
    "    df_exclusive_2023_transcripts[tissue] = tmp.loc[tmp[\"transcript_id\"].isin(exclusive_2023_transcripts[\"transcript_id\"])].copy()\n",
    "\n",
    "    prep_for_exclusive_2023_tx_df = df_exclusive_2023_transcripts[tissue].copy()[['gene_id', 'gene_name', 'transcript_id', 'tissue_site_detail', 'median_CPM', 'median_unique_counts', 'median_relative_abundance', 'median_gene_cpm', 'transcript_biotype', 'gene_biotype']]\n",
    "    exclusive_2023_tx_df = pd.concat([exclusive_2023_tx_df, prep_for_exclusive_2023_tx_df.copy()])\n",
    "    \n",
    "    # looking at only those with a new cds region\n",
    "    df_exclusive_2023_ccds[tissue] = tmp.loc[tmp[\"transcript_id\"].isin(exclusive_2023_cds[\"transcript_id\"])].copy()\n",
    "    \n",
    "    # create empty list for the tissue\n",
    "    list_2023_all_transcript_median[tissue] = []\n",
    "    list_2023_all_gene_median[tissue] = []\n",
    "    list_2023_new_cds_transcript_median[tissue] = []\n",
    "    list_2023_med_relevant_transcript_median[tissue] = []\n",
    "    list_2023_med_relevant_new_cds_transcript_median[tissue] = []\n",
    "    list_2023_brain_relevant_transcript_median[tissue] = []\n",
    "    list_2023_brain_relevant_new_cds_transcript_median[tissue] = []\n",
    "    \n",
    "    # for a range of CPM thresholds\n",
    "    for i in range(0, 1002):\n",
    "\n",
    "        cpm_thresh = i/100\n",
    "\n",
    "        # get the isoforms that fit each category\n",
    "        median_2023 = df_exclusive_2023_transcripts[tissue].loc[df_exclusive_2023_transcripts[tissue][\"median_CPM\"] >= cpm_thresh].copy()\n",
    "        new_cds_median_2023 = df_exclusive_2023_ccds[tissue].loc[df_exclusive_2023_ccds[tissue][\"median_CPM\"] >= cpm_thresh].copy()\n",
    "        med_relevant_median_2023 = median_2023.loc[median_2023[\"gene_id\"].isin(disease_relevant_genes[\"gene_id\"])].copy()\n",
    "        med_relevant_new_cds_median_2023 = med_relevant_median_2023.loc[med_relevant_median_2023[\"transcript_id\"].isin(new_cds_median_2023[\"transcript_id\"])].copy()\n",
    "        brain_relevant_median_2023 = median_2023.loc[median_2023[\"gene_id\"].isin(brain_disease_gene_ids[\"gene_id\"])].copy()\n",
    "        brain_relevant_new_cds_median_2023 = brain_relevant_median_2023.loc[brain_relevant_median_2023[\"transcript_id\"].isin(new_cds_median_2023[\"transcript_id\"])].copy()\n",
    "        \n",
    "        # at CPM > 1 (in this case, because we round values to 2 decimals, CPM == 1.01 is equivalent to CPM > 1)\n",
    "        if cpm_thresh == 1.01:\n",
    "            # grab modified dataframes for later analysis\n",
    "            isoforms_in_categories[tissue] = {\n",
    "                \"All isoforms\": median_2023[['gene_id', 'gene_name', 'transcript_id', 'median_relative_abundance']],\n",
    "                \"Med-relevant\": med_relevant_median_2023[['gene_id', 'gene_name', 'transcript_id', 'median_relative_abundance']],\n",
    "                \"New protein coding sequence\": new_cds_median_2023[['gene_id', 'gene_name', 'transcript_id', 'median_relative_abundance']],\n",
    "                \"Med-relevant new protein coding sequence\": med_relevant_new_cds_median_2023[['gene_id', 'gene_name', 'transcript_id', 'median_relative_abundance']],\n",
    "                \"Brain disease relevant\": brain_relevant_median_2023[['gene_id', 'gene_name', 'transcript_id', 'median_relative_abundance']],\n",
    "                \"Brain disease relevant new protein coding sequence\": brain_relevant_new_cds_median_2023[['gene_id', 'gene_name', 'transcript_id', 'median_relative_abundance']],\n",
    "            }\n",
    "\n",
    "        # add just the total number in each category to the list\n",
    "        list_2023_all_transcript_median[tissue].append(median_2023.shape[0])\n",
    "        list_2023_all_gene_median[tissue].append(median_2023['gene_id'].nunique())\n",
    "        list_2023_new_cds_transcript_median[tissue].append(new_cds_median_2023.shape[0])\n",
    "        list_2023_med_relevant_transcript_median[tissue].append(med_relevant_median_2023.shape[0])\n",
    "        list_2023_med_relevant_new_cds_transcript_median[tissue].append(med_relevant_new_cds_median_2023.shape[0])\n",
    "        list_2023_brain_relevant_transcript_median[tissue].append(brain_relevant_median_2023.shape[0])\n",
    "        list_2023_brain_relevant_new_cds_transcript_median[tissue].append(brain_relevant_new_cds_median_2023.shape[0])\n",
    "       \n",
    "    # concat values to dataframe holding all the numbers for the thresholds for all tissues\n",
    "    values_at_cpm_threshold = pd.concat([values_at_cpm_threshold, pd.DataFrame({\n",
    "        'cpm_threshold': list_cpm_thresh,\n",
    "        'tissue': tissue,\n",
    "        'all_transcript_median': list_2023_all_transcript_median[tissue],\n",
    "        'all_gene_median': list_2023_all_gene_median[tissue],\n",
    "        'new_cds_transcript_median': list_2023_new_cds_transcript_median[tissue],\n",
    "        'med_relevant_transcript_median': list_2023_med_relevant_transcript_median[tissue],\n",
    "        'med_relevant_new_cds_transcript_median': list_2023_med_relevant_new_cds_transcript_median[tissue],\n",
    "        'brain_relevant_transcript_median': list_2023_brain_relevant_transcript_median[tissue],\n",
    "        'brain_relevant_new_cds_transcript_median': list_2023_brain_relevant_new_cds_transcript_median[tissue]\n",
    "    })])\n",
    "\n",
    "\n",
    "# write to files\n",
    "values_at_cpm_threshold.to_csv(\"../../tables/GTEx_expression_since_2019/gtex_values_at_cpm_thresholds_new_since_2019.tsv\", sep = '\\t', index=False)\n",
    "exclusive_2023_tx_df.to_csv('../../tables/GTEx_expression_since_2019/gtex_values_cpm_tx_new_since_2019.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce42992d",
   "metadata": {},
   "source": [
    "## Run venn comparisons between tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c24d9142",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'lung,cereb,front': 8, 'lung,cereb': 3, 'cereb,front': 7, 'front,lung': 3, 'lung': 20, 'cereb': 19, 'front': 5}\n",
      "ALL:\n",
      "{'ENST00000678921', 'ENST00000676712', 'ENST00000677047', 'ENST00000679902', 'ENST00000696840', 'ENST00000707132', 'ENST00000678242', 'ENST00000682710'}\n",
      "just lung\n",
      "{'ENST00000676865', 'ENST00000680422', 'ENST00000677976', 'ENST00000676502', 'ENST00000679172', 'ENST00000678615', 'ENST00000678727', 'ENST00000677207', 'ENST00000698131', 'ENST00000699277', 'ENST00000678886', 'ENST00000679214', 'ENST00000677418', 'ENST00000678031', 'ENST00000679224', 'ENST00000677544', 'ENST00000678929', 'ENST00000679887', 'ENST00000679051', 'ENST00000678114'}\n"
     ]
    }
   ],
   "source": [
    "# compare two brain regions and lung\n",
    "lung_br = set(isoforms_in_categories['Lung']['Brain disease relevant']['transcript_id'].tolist())\n",
    "cereb_br = set(isoforms_in_categories['Brain - Cerebellar Hemisphere']['Brain disease relevant']['transcript_id'].tolist())\n",
    "front_br = set(isoforms_in_categories['Brain - Frontal Cortex (BA9)']['Brain disease relevant']['transcript_id'].tolist())\n",
    "\n",
    "# Get the full intersect of 'transcript_id' from all DataFrames\n",
    "full_intersect = lung_br & cereb_br & front_br\n",
    "\n",
    "# Get the intersect of 'transcript_id' between two DataFrames\n",
    "lung_and_cereb = lung_br & cereb_br\n",
    "cereb_and_front = cereb_br & front_br\n",
    "front_and_lung = front_br & lung_br\n",
    "\n",
    "# Calculate the lengths for each set operation\n",
    "overlap_lengths = {\n",
    "    'lung,cereb,front': len(full_intersect),\n",
    "    'lung,cereb': len(lung_and_cereb - full_intersect),\n",
    "    'cereb,front': len(cereb_and_front - full_intersect),\n",
    "    'front,lung': len(front_and_lung - full_intersect),\n",
    "    'lung': len(lung_br - (lung_and_cereb | front_and_lung)),\n",
    "    'cereb': len(cereb_br - (lung_and_cereb | cereb_and_front)),\n",
    "    'front': len(front_br - (cereb_and_front | front_and_lung))\n",
    "}\n",
    "\n",
    "print(overlap_lengths)\n",
    "print(\"ALL:\")\n",
    "print(full_intersect)\n",
    "print(\"just lung\")\n",
    "print(lung_br - (lung_and_cereb | front_and_lung))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "78d3e971",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ENST00000678921', 'ENST00000678242', 'ENST00000679902'}\n"
     ]
    }
   ],
   "source": [
    "# look at the intersection of all \n",
    "put_br = set(isoforms_in_categories['Brain - Putamen (basal ganglia)']['Brain disease relevant']['transcript_id'].tolist())\n",
    "cereb_br = set(isoforms_in_categories['Brain - Cerebellar Hemisphere']['Brain disease relevant']['transcript_id'].tolist())\n",
    "front_br = set(isoforms_in_categories['Brain - Frontal Cortex (BA9)']['Brain disease relevant']['transcript_id'].tolist())\n",
    "cells_br = set(isoforms_in_categories['Cells - Cultured fibroblasts']['Brain disease relevant']['transcript_id'].tolist())\n",
    "haa_br = set(isoforms_in_categories['Heart - Atrial Appendage']['Brain disease relevant']['transcript_id'].tolist())\n",
    "hlv_br = set(isoforms_in_categories['Heart - Left Ventricle']['Brain disease relevant']['transcript_id'].tolist())\n",
    "lung_br = set(isoforms_in_categories['Lung']['Brain disease relevant']['transcript_id'].tolist())\n",
    "liv_br = set(isoforms_in_categories['Liver']['Brain disease relevant']['transcript_id'].tolist())\n",
    "mus_br = set(isoforms_in_categories['Muscle - Skeletal']['Brain disease relevant']['transcript_id'].tolist())\n",
    "\n",
    "print(put_br & cereb_br & front_br & cells_br & haa_br & hlv_br & lung_br & liv_br & mus_br)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9c468b8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ENST00000678921', 'ENST00000674619', 'ENST00000696840', 'ENST00000679902', 'ENST00000707132', 'ENST00000678242', 'ENST00000675737', 'ENST00000676437'}\n"
     ]
    }
   ],
   "source": [
    "# look at the intersection of brain regions\n",
    "print(put_br & cereb_br & front_br)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "85f050d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'lung,cereb,cells': 8, 'lung,cereb': 3, 'cereb,cells': 1, 'cells,lung': 9, 'lung': 14, 'cereb': 25, 'cells': 6}\n",
      "ALL:\n",
      "{'ENST00000678921', 'ENST00000676712', 'ENST00000677047', 'ENST00000679902', 'ENST00000696840', 'ENST00000707132', 'ENST00000678242', 'ENST00000682710'}\n",
      "just lung:\n",
      "{'ENST00000676865', 'ENST00000680422', 'ENST00000677976', 'ENST00000698131', 'ENST00000679172', 'ENST00000699277', 'ENST00000678886', 'ENST00000678114', 'ENST00000678727', 'ENST00000678031', 'ENST00000679224', 'ENST00000676808', 'ENST00000677207', 'ENST00000679887'}\n"
     ]
    }
   ],
   "source": [
    "# compare lung, cerebellum, and cultured fibroblasts\n",
    "lung_br = set(isoforms_in_categories['Lung']['Brain disease relevant']['transcript_id'].tolist())\n",
    "cereb_br = set(isoforms_in_categories['Brain - Cerebellar Hemisphere']['Brain disease relevant']['transcript_id'].tolist())\n",
    "cell_br = set(isoforms_in_categories['Cells - Cultured fibroblasts']['Brain disease relevant']['transcript_id'].tolist())\n",
    "\n",
    "# Get the full intersect of 'transcript_id' from all DataFrames\n",
    "full_intersect = lung_br & cereb_br & front_br\n",
    "\n",
    "# Get the intersect of 'transcript_id' between two DataFrames\n",
    "lung_and_cereb = lung_br & cereb_br\n",
    "cereb_and_cell = cereb_br & cell_br\n",
    "cell_and_lung = cell_br & lung_br\n",
    "\n",
    "# Calculate the lengths for each set operation\n",
    "overlap_lengths = {\n",
    "    'lung,cereb,cells': len(full_intersect),\n",
    "    'lung,cereb': len(lung_and_cereb - full_intersect),\n",
    "    'cereb,cells': len(cereb_and_cell - full_intersect),\n",
    "    'cells,lung': len(cell_and_lung - full_intersect),\n",
    "    'lung': len(lung_br - (lung_and_cereb | cell_and_lung)),\n",
    "    'cereb': len(cereb_br - (lung_and_cereb | cereb_and_cell)),\n",
    "    'cells': len(cell_br - (cereb_and_cell | cell_and_lung))\n",
    "}\n",
    "\n",
    "print(overlap_lengths)\n",
    "print(\"ALL:\")\n",
    "print(full_intersect)\n",
    "print('just lung:')\n",
    "print(lung_br - (lung_and_cereb | cell_and_lung))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e072af83",
   "metadata": {},
   "source": [
    "## Look at relative expression of new isoforms "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c4c50b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create dataframe to store percent expression\n",
    "percent_expression_df = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4bbb8fd4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# gene/isoform categories\n",
    "cat_to_df = {\"Med-relevant\", \"New protein coding sequence\", \"Med-relevant new protein coding sequence\", \"Brain disease relevant\", \"Brain disease relevant new protein coding sequence\"}\n",
    "\n",
    "\n",
    "for tissue in tissues_to_use:\n",
    "    for category in cat_to_df:\n",
    "        ## Get percent expression from transcripts since 2019 for each subcategory\n",
    "        df_tmp = isoforms_in_categories[tissue][category].copy()\n",
    "        df_tmp_comb_rel_exp = df_tmp.groupby(['gene_id', 'gene_name'])['median_relative_abundance'].sum().reset_index()\n",
    "        df_tmp_comb_rel_exp.rename(columns={'median_relative_abundance': 'Relative Abundance of New Transcripts Since 2019 (%)'}, inplace=True)\n",
    "        df_tmp_comb_rel_exp['Category'] = category\n",
    "        df_tmp_comb_rel_exp['Tissue'] = tissue\n",
    "        percent_expression_df = pd.concat([percent_expression_df, df_tmp_comb_rel_exp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e9784013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Relative Abundance of New Transcripts Since 2019 (%)</th>\n",
       "      <th>Category</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000001497</td>\n",
       "      <td>LAS1L</td>\n",
       "      <td>3.420882</td>\n",
       "      <td>Med-relevant new protein coding sequence</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000003393</td>\n",
       "      <td>ALS2</td>\n",
       "      <td>3.611826</td>\n",
       "      <td>Med-relevant new protein coding sequence</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004455</td>\n",
       "      <td>AK2</td>\n",
       "      <td>48.969782</td>\n",
       "      <td>Med-relevant new protein coding sequence</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000008441</td>\n",
       "      <td>NFIX</td>\n",
       "      <td>0.398749</td>\n",
       "      <td>Med-relevant new protein coding sequence</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000021645</td>\n",
       "      <td>NRXN3</td>\n",
       "      <td>0.511729</td>\n",
       "      <td>Med-relevant new protein coding sequence</td>\n",
       "      <td>Brain - Cerebellar Hemisphere</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name  \\\n",
       "0  ENSG00000001497     LAS1L   \n",
       "1  ENSG00000003393      ALS2   \n",
       "2  ENSG00000004455       AK2   \n",
       "3  ENSG00000008441      NFIX   \n",
       "4  ENSG00000021645     NRXN3   \n",
       "\n",
       "   Relative Abundance of New Transcripts Since 2019 (%)  \\\n",
       "0                                              3.420882   \n",
       "1                                              3.611826   \n",
       "2                                             48.969782   \n",
       "3                                              0.398749   \n",
       "4                                              0.511729   \n",
       "\n",
       "                                   Category                         Tissue  \n",
       "0  Med-relevant new protein coding sequence  Brain - Cerebellar Hemisphere  \n",
       "1  Med-relevant new protein coding sequence  Brain - Cerebellar Hemisphere  \n",
       "2  Med-relevant new protein coding sequence  Brain - Cerebellar Hemisphere  \n",
       "3  Med-relevant new protein coding sequence  Brain - Cerebellar Hemisphere  \n",
       "4  Med-relevant new protein coding sequence  Brain - Cerebellar Hemisphere  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# write to file to use in plots\n",
    "percent_expression_df.to_csv(\"../../tables/GTEx_expression_since_2019/gtex_cpm_gt_1_relative_abundance_new_since_2019.tsv\", sep = '\\t', index=False)\n",
    "percent_expression_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c6fd77e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
